

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Viewing COS Data &#8212; HST Notebooks</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/table.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/COS/ViewData/ViewData';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Modifying or Creating an Association File" href="../AsnFile/AsnFile.html" />
    <link rel="prev" title="Downloading COS Data" href="../DataDl/DataDl.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/stsci_logo2.png" class="logo__image only-light" alt="HST Notebooks - Home"/>
    <script>document.write(`<img src="../../../_static/stsci_logo2.png" class="logo__image only-dark" alt="HST Notebooks - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../index.html">
                    STScI HST Notebook Repository HQ
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">ACS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../ACS/README.html">ACS Notebooks</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_findsat_mrt/acs_findsat_mrt_example.html">Satellite trail detection in ACS/WFC data using acstools.findsat_mrt</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_saturation_trails/acs_saturation_trails.html">ACS Linearity with Saturated Stars</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_cte_forward_model/acs_cte_forward_model_example.html">Pixel-based ACS/WFC CTE Forward Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_subarrays/acs_subarrays.html">ACS Image Reduction for Subarray Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_sbc_dark_analysis/acs_sbc_dark_analysis.html">SBC Dark Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_reduction/acs_reduction.html">ACS/WFC Image Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_pixel_area_maps/acs_pixel_area_maps.html">Obtaining Pixel Area Maps for ACS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_polarization_tools/acs_polarization_tools.html">Using ACS Polarization Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_focus_diverse_epsfs/acs_focus_diverse_epsfs.html">Focus Diverse ePSFs for ACS/WFC</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_exception_report/exception_report_checks.html">HST Exception Report - Investigate your ACS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/hst_orbits_ephem/hst_orbits_ephem.html">Getting HST Ephemeris and Related Data for an Observation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">COS</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../README.html">COS Notebooks</a></li>



<li class="toctree-l1"><a class="reference internal" href="../Setup/Setup.html">Setting up your computer environment for working with COS data</a></li>





<li class="toctree-l1"><a class="reference internal" href="../DataDl/DataDl.html">Downloading COS Data</a></li>




<li class="toctree-l1 current active"><a class="current reference internal" href="#">Viewing COS Data</a></li>





<li class="toctree-l1"><a class="reference internal" href="../AsnFile/AsnFile.html">Modifying or Creating an Association File</a></li>





<li class="toctree-l1"><a class="reference internal" href="../CalCOS/CalCOS.html">Running the COS Data Pipeline (<code class="docutils literal notranslate"><span class="pre">CalCOS</span></code>)</a></li>






<li class="toctree-l1"><a class="reference internal" href="../SplitTag/SplitTag.html">Splitting COS Exposures into sub-exposures with <code class="docutils literal notranslate"><span class="pre">splittag</span></code></a></li>







<li class="toctree-l1"><a class="reference internal" href="../DayNight/DayNight.html">Filtering out COS Data taken during the Day or Night</a></li>
<li class="toctree-l1"><a class="reference internal" href="../LSF/LSF.html">Working with the COS Line Spread Function (LSF)</a></li>

<li class="toctree-l1"><a class="reference internal" href="../Extract/Extract.html">Editing the extraction boxes in a spectral extraction file (<code class="docutils literal notranslate"><span class="pre">XTRACTAB</span></code>)</a></li>







<li class="toctree-l1"><a class="reference internal" href="../ExceptionReport/ExceptionReport.html">What to do when you get an Exception Report for COS</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">DrizzlePac</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/README.html">DrizzlePac Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/using_updated_astrometry_solutions/using_updated_astrometry_solutions.html">Improving Astrometry Using Alternate WCS Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_to_catalogs/align_to_catalogs.html">Aligning HST images to an Absolute Reference Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_sparse_fields/align_sparse_fields.html">Aligning Deep Exposures of Sparse Fields</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_multiple_visits/align_multiple_visits.html">Optimizing Image Alignment for Multiple HST Visits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/use_ds9_regions_in_tweakreg/use_ds9_regions_in_tweakreg.html">TWEAKREG Tips - Using DS9 Regions for Source Inclusion/Exclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/optimize_image_sampling/optimize_image_sampling.html">Optimizing the Image Sampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_mosaics/align_mosaics.html">Aligning HST Mosaics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/sky_matching/sky_matching.html">Sky Matching for HST Mosaics <a id="top"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/mask_satellite/mask_satellite.html">Satellite Trail Masking Techniques</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/drizzle_wfpc2/drizzle_wfpc2.html">Drizzling new WFPC2 FLT data products with chip-normalization   </a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">HASP</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../HASP/Setup/Setup.html">Installing the Hubble Advanced Spectral Products Script</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../HASP/CoaddTutorial/CoaddTutorial.html">Inputting User Data using the Hubble Advanced Spectral Products Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../HASP/FluxScaleTutorial/FluxScaleTutorial.html">Scaling Flux while using the Hubble Advanced Spectral Products Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../HASP/DataDiagnostic/DataDiagnostics.html">HASP Data Diagnostic Notebook</a></li>




<li class="toctree-l1"><a class="reference internal" href="../../HASP/WavelengthAdjustment/WavelengthAdjustment.html">Wavelength Adjustments for Running the Hubble Advanced Spectral Products (HASP) Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../HASP/COSLifetimePositions/COSLifetimePositions.html">Combining COS Data from Multiple Lifetime Positions and Central Wavelengths</a></li>

</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">NICMOS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../NICMOS/nicmos_unit_conversion/nicmos_unit_conversion.html">Programmatic Replacements for NICMOS Units Conversion Form</a></li>


</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">STIS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../STIS/README.html">STIS-Notebooks</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../STIS/cross-correlation/cross-correlation.html">Correcting for Missing Wavecals with Cross-Correlation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/custom_ccd_darks/custom_ccd_darks.html">Custom CCD Darks <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/drizpac_notebook/STIS_DrizzlePac_Tutorial.html">STIS DrizzlePac Tutorial <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/CoronagraphyViz/STIS_Coronagraphy_Visualization_v2.html">STIS Coronagraphy Visualization Tool (v2)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/calstis/calstis_2d_ccd.html">Calstis 2-D CCD Data Reduction <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/target_acquisition/target_acquisition.html">Evaluating STIS Target Acquisitions <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/view_data/view_data.html">Viewing STIS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/extraction/1D_Extraction.html">1D Spectra Extraction <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/low_count_uncertainties/Low_Count_Uncertainties.html">Low Count Uncertainties in STIS <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/contrast_sensitivity/STIS_Coronagraphic_Observation_Feasibility.html">STIS Coronagraphic Observation Feasibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">WFC3</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../WFC3/README.html">WFC3 Notebooks</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/general_tools.html">General Tools</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/image_displayer_analyzer/wfc3_image_displayer_analyzer.html">WFC3 Image Displayer &amp; Analyzer</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/exception_report/wfc3_exception_report.html">Exception Report Checklist - WFC3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/calwf3_v1.0_cte/calwf3_with_v1.0_PCTE.html">Processing WFC3/UVIS Data with <code class="docutils literal notranslate"><span class="pre">calwf3</span></code> Using the v1.0 CTE-Correction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/persistence/wfc3_ir_persistence.html">Masking Persistence in WFC3/IR Images</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_g280_transit/G280_Exoplanet_Transits.html">Analyzing WFC3/UVIS G280 Exoplanet Transit Observations</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/ir_tvb.html">WFC3/IR Time Variable Background (TVB)</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.html">WFC3/IR IMA Visualization Tools with An Example of Time Variable Background</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/calwf3_recalibration/calwf3_recal_tvb.html">Manual Recalibration with calwf3: Turning off the IR Linear Ramp Fit</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/tvb_flattenramp/TVB_flattenramp_notebook.html">Correcting for Helium Line Emission Background in IR Exposures using the “Flatten-Ramp” Technique</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_scattered_light_calwf3_corrections/Correcting_for_Scattered_Light_in_IR_Exposures_Using_calwf3_to_Mask_Bad_Reads.html">Correcting for Scattered Light in WFC3/IR Exposures: Using <code class="docutils literal notranslate"><span class="pre">calwf3</span></code> to Mask Bad Reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_scattered_light_manual_corrections/Correcting_for_Scattered_Light_in_IR_Exposures_by_Manually_Subtracting_Bad_Reads.html">Correcting for Scattered Light in WFC3/IR Exposures: Manually Subtracting Bad Reads</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/photometry.html">Photometry</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/filter_transformations/filter_transformations.html">WFC3/UVIS Filter Transformations with stsynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/flux_conversion_tool/flux_conversion_tool.html">Flux Unit Conversions with synphot and stsynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/photometry_examples/phot_examples.html">Synthetic Photometry Examples for WFC3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_time_dependent_photometry/uvis_timedep_phot.html">WFC3/UVIS Time-dependent Photometry</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/zeropoints/zeropoints.html">Calculating WFC3 Zeropoints with STSynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_pam_corrections/WFC3_UVIS_Pixel_Area_Map_Corrections_for_Subarrays.html">WFC3/UVIS Pixel Area Map Corrections for Subarrays</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/point_spread_function.html">Point Spread Function (PSF)</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/point_spread_function/hst_point_spread_function.html">HST WFC3 Point Spread Function Modeling   </a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/mast_api_psf/download_psf_cutouts.html">Downloading WFC3 and WFPC2 PSF Cutouts from MAST</a></li>
</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/spacetelescope/hst_notebooks/blob/main/notebooks/COS/ViewData/ViewData.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="../../../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="initThebeSBT()"
  class="btn btn-sm btn-launch-thebe dropdown-item"
  title="Launch Thebe"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-play"></i>
  </span>
<span class="btn__text-container">Live Code</span>
</button>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/spacetelescope/hst_notebooks" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/spacetelescope/hst_notebooks/issues/new?title=Issue%20on%20page%20%2Fnotebooks/COS/ViewData/ViewData.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/notebooks/COS/ViewData/ViewData.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Viewing COS Data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Viewing COS Data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning Goals</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">0. Introduction</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-one-cell-summary-of-this-notebook-s-key-points">0.1. A one-cell summary of this Notebook’s key points:</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#now-returning-to-our-more-detailed-walkthrough">Now, returning to our more detailed walkthrough</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#notes-for-those-new-to-python-jupyter-coding">Notes for those new to <code class="docutils literal notranslate"><span class="pre">Python</span></code>/Jupyter/Coding:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-in-the-data">1. Reading in the data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#investigating-the-data-basics">1.1. Investigating the Data - <em>Basics</em></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-in-the-x1d-x1dsum-main-data">1.2. Reading in the <code class="docutils literal notranslate"><span class="pre">x1d</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> Main Data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fig-1-1-from-cos-dhb-fig-1-6">Fig. 1.1 from COS DHB Fig. 1.6</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fig-1-2-from-cos-dhb-fig-1-10">Fig. 1.2 from COS DHB Fig. 1.10</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-association-asn-file">1.3. The Association (<code class="docutils literal notranslate"><span class="pre">asn</span></code>) file</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1-1-finding-metadata-for-the-nuv">Exercise 1.1. <em>Finding Metadata for the NUV</em></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-data">2. Plotting the Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-an-fuv-spectrum">2.1 Plotting an FUV Spectrum</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-first-plot">2.1.1 Our First Plot</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-1-a-clearer-plot">Exercise 2.1: A clearer plot</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-complex-look-at-the-entire-fuv">2.1.2. A Complex Look at the Entire FUV</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-closer-at-parts-of-the-fuv-spectrum">2.1.3. Looking Closer at Parts of the FUV Spectrum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-2-choosing-and-plotting-a-useful-data-range">Exercise 2.2 Choosing and plotting a useful data range</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-plotting-the-data-with-specutils-optional">2.1.4. Reading and plotting the data with <code class="docutils literal notranslate"><span class="pre">specutils</span></code> (<em>OPTIONAL</em>)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-an-nuv-spectrum">2.2. Plotting an NUV Spectrum</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examining-the-first-order-spectrum">2.2.1. Examining the first-order spectrum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examining-the-second-order-spectrum">2.2.2. Examining the second-order spectrum</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#preliminary-review-of-the-data">3. Preliminary Review of the Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-quality-and-initial-cleanup">3.1. Data Quality and Initial Cleanup</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-1-removing-all-flagged-data-from-the-nuv">Exercise 3.1: <em>Removing all flagged data from the NUV</em></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#binning">3.2. Binning</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bringing-in-some-useful-functions-for-binning">3.2.1. Bringing in Some Useful Functions for Binning</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#binning-the-fuv-data">3.2.2. Binning the FUV Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#estimating-the-signal-to-noise-snr-ratio">3.3. Estimating the Signal-to-Noise (SNR) Ratio</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bringing-in-a-useful-function-for-estimating-snr">3.3.1.  Bringing in a Useful Function for Estimating SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-a-region-to-calculate-snr">3.3.2. Choosing a Region to calculate SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-the-snr">3.3.3. Calculating the SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-2-binning-the-nuv-data-and-calculating-its-snr">Exercise 3.2. <em>Binning the NUV Data and Calculating its SNR</em></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#congratulations-you-finished-this-notebook">Congratulations! You finished this Notebook!</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#there-are-more-cos-data-walkthrough-notebooks-on-different-topics-you-can-find-them-here">There are more COS data walkthrough Notebooks on different topics. You can find them here.</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#about-this-notebook">About this Notebook</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citations">Citations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-solutions">Exercise Solutions:</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <p><a id="topV"></a></p>
<section class="tex2jax_ignore mathjax_ignore" id="viewing-cos-data">
<h1>Viewing COS Data<a class="headerlink" href="#viewing-cos-data" title="Permalink to this heading">#</a></h1>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="learning-goals">
<h1>Learning Goals<a class="headerlink" href="#learning-goals" title="Permalink to this heading">#</a></h1>
<p><font size="4 ">This Notebook is designed to walk the user (<em>you</em>) through:</font><br></p>
<p><strong>0. <span class="xref myst">Introduction</span></strong></p>
<p>- 0.1. <span class="xref myst">A one-cell summary of this Notebook’s key points</span></p>
<p><strong>1. <span class="xref myst">Reading in the data</span> using <code class="docutils literal notranslate"><span class="pre">Python</span></code></strong></p>
<p>- 1.1. <span class="xref myst">Investigating the Data - <em>Basics</em></span></p>
<p>- 1.2. <span class="xref myst">Reading in the <code class="docutils literal notranslate"><span class="pre">x1d</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> Main Data</span></p>
<p>- 1.3. <span class="xref myst">The Association (<code class="docutils literal notranslate"><span class="pre">asn</span></code>) file</span></p>
<p><strong>2. <span class="xref myst">Displaying the data</span> using common plotting techniques</strong></p>
<p>- 2.1. <span class="xref myst">Plotting an FUV Spectrum</span></p>
<blockquote>
<div><p>2.1.1. <span class="xref myst">Our First Plot</span></p>
</div></blockquote>
<blockquote>
<div><p>2.1.2. <span class="xref myst">A Complex Look at the Entire FUV</span></p>
</div></blockquote>
<blockquote>
<div><p>2.1.3. <span class="xref myst">Looking Closer at Parts of the FUV Spectrum</span></p>
</div></blockquote>
<blockquote>
<div><p>2.1.4. <span class="xref myst">Reading and plotting the data with <code class="docutils literal notranslate"><span class="pre">specutils</span></code> (<em>optional</em>)</span></p>
</div></blockquote>
<p>- 2.2. <span class="xref myst">Plotting an NUV Spectrum</span></p>
<blockquote>
<div><p>2.2.1. <span class="xref myst">Examining the first-order spectrum</span></p>
</div></blockquote>
<blockquote>
<div><p>2.2.2. <span class="xref myst">Examining the second-order spectrum</span></p>
</div></blockquote>
<p><strong>3. <span class="xref myst">Making a quick assessment of the data</span> with tools to bin and measure the data’s SNR and resolution</strong></p>
<p>- 3.1. <span class="xref myst">Understanding and Using Data Quality Flags</span></p>
<p>- 3.2. <span class="xref myst">Binning the Data</span></p>
<blockquote>
<div><p>3.2.1. <span class="xref myst">Bringing in Some Useful Functions for Binning</span></p>
</div></blockquote>
<blockquote>
<div><p>3.2.2. <span class="xref myst">Binning the FUV Data</span></p>
</div></blockquote>
<p>- 3.3. <span class="xref myst">Calculating the Signal-to-Noise Ratio</span></p>
<blockquote>
<div><p>3.3.1. <span class="xref myst">Defining a Useful Function for Estimating SNR</span></p>
</div></blockquote>
<blockquote>
<div><p>3.3.2. <span class="xref myst">Choosing a Region to calculate SNR</span></p>
</div></blockquote>
<blockquote>
<div><p>3.3.3. <span class="xref myst">Calculating the SNR</span></p>
</div></blockquote>
<p><em>If you’re confident with Reading-in and Plotting data in <code class="docutils literal notranslate"><span class="pre">Python</span></code>, you may run the first few cells defining directories and importing modules and then click the link when it appears to go ahead and skip to <span class="xref myst">Section 3</span></em></p>
<p><a id=introV></a></p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="introduction">
<h1>0. Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">#</a></h1>
<p><strong>The Cosmic Origins Spectrograph (<a class="reference external" href="https://www.nasa.gov/content/hubble-space-telescope-cosmic-origins-spectrograph"><em>COS</em></a>) is an ultraviolet spectrograph on-board the Hubble Space Telescope (<a class="reference external" href="https://www.stsci.edu/hst/about"><em>HST</em></a>) with capabilities in the near ultraviolet (<em>NUV</em>) and far ultraviolet (<em>FUV</em>).</strong></p>
<p><strong>This tutorial aims to prepare you to begin analyzing COS data of your choice by walking you through reading and viewing a spectrum obtained with COS, as well as obtaining a few diagnostic measurements of your spectrum.</strong></p>
<ul class="simple">
<li><p>For an in-depth manual to working with COS data and a discussion of caveats and user tips, see the <a class="reference external" href="https://hst-docs.stsci.edu/display/COSDHB/">COS Data Handbook</a>.</p></li>
<li><p>For a detailed overview of the COS instrument, see the <a class="reference external" href="https://hst-docs.stsci.edu/display/COSIHB/">COS Instrument Handbook</a>.</p></li>
</ul>
<p><a id=onecellV></a></p>
<section id="a-one-cell-summary-of-this-notebook-s-key-points">
<h2>0.1. A one-cell summary of this Notebook’s key points:<a class="headerlink" href="#a-one-cell-summary-of-this-notebook-s-key-points" title="Permalink to this heading">#</a></h2>
<p>While the rest of this Notebook will walk you through each step and decision made when examining your COS data, the following code cell serves as a summary for the Notebook. It contains the key material condensed into a single code cell, without much explanation. If this is all the help you need, great! If you still have questions, read on!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This cell condenses the key material of the Notebook into a single cell</span>

<span class="c1"># 1. Import the necessary libraries:</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.table</span><span class="w"> </span><span class="kn">import</span> <span class="n">Table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">fits</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astroquery.mast</span><span class="w"> </span><span class="kn">import</span> <span class="n">Observations</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cos_functions</span><span class="w"> </span><span class="kn">import</span> <span class="n">estimate_snr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="c1"># 2. Download an example dataset using astroquery:</span>
<span class="c1"># For more information, see the &quot;Downloading your data&quot; Notebook</span>
<span class="n">onecell_x1dsum_dataprod</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
    <span class="c1"># Filtering through data products, only downloading X1DSUM files.</span>
    <span class="n">Observations</span><span class="o">.</span><span class="n">filter_products</span><span class="p">(</span>
        <span class="c1"># Getting a list of data products for the given observation</span>
        <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
            <span class="c1"># Searching for objects in MAST</span>
            <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
                <span class="c1"># The Obs ID of the observation to download</span>
                <span class="n">obs_id</span><span class="o">=</span><span class="s2">&quot;LDM701020&quot;</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="c1"># Only downloads the 1 dimensional extracted spectrum</span>
        <span class="n">productSubGroupDescription</span><span class="o">=</span><span class="s2">&quot;X1DSUM&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="c1"># 3. Read in the data to an Astropy Table:</span>
<span class="n">x1dsum_data_table</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
                        <span class="n">onecell_x1dsum_dataprod</span><span class="p">[</span><span class="s2">&quot;Local Path&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>

<span class="c1"># Some users may be familiar with another way to read FITS table data:</span>
<span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">onecell_x1dsum_dataprod</span><span class="p">[</span><span class="s2">&quot;Local Path&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">hdulist</span><span class="p">:</span>
    <span class="n">alternate_x1dsum_data_structure</span> <span class="o">=</span> <span class="n">hdulist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

<span class="c1"># 4. Estimate the SNR of the spectrum by calculating sqrt(counts):</span>
<span class="c1">#    The `estimate_snr` function is defined `cos_functions.py`</span>
<span class="n">SNR_estimate</span> <span class="o">=</span> <span class="n">estimate_snr</span><span class="p">(</span>
                    <span class="n">x1dsum_data_table</span><span class="p">,</span>
                    <span class="n">binsize_</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
                    <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># 5. Create a plot of the spectrum and the estimated S/N:</span>
<span class="c1"># 2 subplots, populate the top one with flux and the lower with S/N:</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span>
                               <span class="n">sharex</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                               <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">segment_row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">x1dsum_data_table</span><span class="p">):</span>
    <span class="c1"># In the upper panel, plot the spectrum as Flux over Wavelength:</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">segment_row</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">],</span> <span class="n">segment_row</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">],</span>
             <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Segment FUV</span><span class="si">{</span><span class="s1">&#39;AB&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># In the lower panel, plot SNR as estimated S/N over Wavelength:</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">SNR_estimate</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">SNR_estimate</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
             <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Segment FUV</span><span class="si">{</span><span class="s1">&#39;AB&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># The next lines will format the plot and save</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Adds a title of fontsize 20 points</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;Fig. 0.1: Simple COS G160M Spectrum&quot;</span><span class="p">,</span>
              <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="c1"># Adds y axis label to the top panel</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
               <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="c1"># Adds y axis label to the bottom panel</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Estimated SNR for a res element&#39;</span><span class="p">,</span>
               <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="c1"># Adds x axis label to the bottom panel</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
               <span class="n">size</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>

<span class="c1"># Removing whitespace in figure</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="c1"># Saving figure as PNG</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;./Fig_0.1_quick_plot.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/ldm701020_x1dsum.fits to ./mastDownload/HST/ldm701020/ldm701020_x1dsum.fits ...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [Done]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
</pre></div>
</div>
<img alt="../../../_images/e4da7cdee34f67c888414bf63193069dbb67980884be4a3f678047199c58e75c.png" src="../../../_images/e4da7cdee34f67c888414bf63193069dbb67980884be4a3f678047199c58e75c.png" />
</div>
</div>
<section id="now-returning-to-our-more-detailed-walkthrough">
<h3>Now, returning to our more detailed walkthrough<a class="headerlink" href="#now-returning-to-our-more-detailed-walkthrough" title="Permalink to this heading">#</a></h3>
<section id="notes-for-those-new-to-python-jupyter-coding">
<h4>Notes for those new to <code class="docutils literal notranslate"><span class="pre">Python</span></code>/Jupyter/Coding:<a class="headerlink" href="#notes-for-those-new-to-python-jupyter-coding" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p>You will frequently see exclamation points (<strong>!</strong>) or dollar signs (<strong>$</strong>) at the beginning of a line of code we are telling you to run. These are not part of the actual commands. The exclamation points tell a Jupyter Notebook to pass the following line to the command line, and the dollar sign merely indicates the start of a terminal prompt.</p></li>
</ul>
<p><font size="4"> We will import the following packages:</font></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">numpy</span></code> to handle array functions (version <span class="math notranslate nohighlight">\(\ge\)</span> 1.17)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pathlib</span></code> for managing system paths.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astropy.io</span> <span class="pre">fits</span></code> for accessing FITS files</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astropy.table</span> <span class="pre">Table</span></code> for creating tidy tables of the data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astropy.units</span></code> and <code class="docutils literal notranslate"><span class="pre">astropy.visualization.quantity_support</span></code> for dealing with units</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">matplotlib.pyplot</span></code> for plotting data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astroquery.mast</span> <span class="pre">Observations</span></code> for finding and downloading data from the <a class="reference external" href="https://mast.stsci.edu/portal/Mashup/Clients/Mast/Portal.html">MAST</a> archive</p></li>
</ul>
<p>The line <code class="docutils literal notranslate"><span class="pre">%matplotlib</span> <span class="pre">inline</span></code> causes <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> plots to appear in the Notebook instead of possibly showing up in separate windows. Additionally, <code class="docutils literal notranslate"><span class="pre">quantity_support()</span></code> supports plotting <code class="docutils literal notranslate"><span class="pre">astropy</span></code> units in <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code>.</p>
<p><em>Later on, we will import some functions from a local file</em>, <code class="docutils literal notranslate"><span class="pre">cos_functions.py</span></code>, which is installed as part of the same GitHub repository as this Notebook. If you do not see this <code class="docutils literal notranslate"><span class="pre">Python</span></code> file in this directory, you can find it <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/COS/ViewData">here</a>. <code class="docutils literal notranslate"><span class="pre">cos_functions.py</span></code> must be installed next to this Notebook (<code class="docutils literal notranslate"><span class="pre">ViewData.ipynb</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">astroquery.mast</span><span class="w"> </span><span class="kn">import</span> <span class="n">Observations</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.table</span><span class="w"> </span><span class="kn">import</span> <span class="n">Table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">fits</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.visualization</span><span class="w"> </span><span class="kn">import</span> <span class="n">quantity_support</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy</span><span class="w"> </span><span class="kn">import</span> <span class="n">units</span> <span class="k">as</span> <span class="n">u</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">quantity_support</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;astropy.visualization.units.quantity_support.&lt;locals&gt;.MplQuantityConverter at 0x7fc1dff84490&gt;
</pre></div>
</div>
</div>
</div>
<p><font size="4"> We will also define a few directories in which to place our data and plots, as well as a few colors we will use in plots later.</font></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># These are the paths for important directories for this Notebook</span>
<span class="n">datadir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;./data&#39;</span><span class="p">)</span>
<span class="n">outputsdir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;./output/&#39;</span><span class="p">)</span>
<span class="n">plotsdir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;./output/plots&#39;</span><span class="p">)</span>

<span class="c1"># Make the directories if they don&#39;t exist</span>
<span class="n">datadir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">outputsdir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plotsdir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Specifying a few colors to correspond to COS segments for plotting</span>
<span class="c1"># by their hex code. Many search engines will show you the hex color</span>
<span class="n">segment_colors</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;FUVA&#39;</span><span class="p">:</span> <span class="s1">&#39;#BC8C5B&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;FUVB&#39;</span><span class="p">:</span> <span class="s1">&#39;#4B6CA4&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;NUVA&#39;</span><span class="p">:</span> <span class="s1">&#39;#1813CE&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;NUVB&#39;</span><span class="p">:</span> <span class="s1">&#39;#61946E&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;NUVC&#39;</span><span class="p">:</span> <span class="s1">&#39;#8C1A11&#39;</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p><font size="4">We will be working with some FUV and NUV datasets downloaded in the cell below. </font></p>
<p>These two datasets contain FUV observations of the QSO <em>3C48</em> and NUV observations of the Star <em>WD1057 + 719</em>, respectively.</p>
<p>Searching for and downloading data is out of the scope of this tutorial. If you wish to learn more, please see our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/blob/main/notebooks/COS/DataDl/DataDl.ipynb">tutorial on downloading COS data</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Download the _x1dsum and _asn NUV data on WD1057+719 (G230L grating)</span>
<span class="n">nuv_downloads</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
                        <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
                            <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
                                <span class="n">obs_id</span><span class="o">=</span><span class="s1">&#39;lbbd01020&#39;</span><span class="p">)),</span>
                        <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">datadir</span><span class="p">),</span>
                        <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;fits&#39;</span><span class="p">,</span>
                        <span class="n">mrp_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">cache</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Download the _x1dsum and _asn FUV data on QSO 3C48</span>
<span class="n">fuv_downloads</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
                        <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
                            <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
                                <span class="n">obs_id</span><span class="o">=</span><span class="s1">&#39;lcxv13050&#39;</span><span class="p">)),</span>
                        <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">datadir</span><span class="p">),</span>
                        <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;fits&#39;</span><span class="p">,</span>
                        <span class="n">mrp_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">cache</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/lbbd01020_asn.fits to data/mastDownload/HST/lbbd01020/lbbd01020_asn.fits ...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [Done]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/lbbd01020_x1dsum.fits to data/mastDownload/HST/lbbd01020/lbbd01020_x1dsum.fits ...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [Done]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/lcxv13050_asn.fits to data/mastDownload/HST/lcxv13050/lcxv13050_asn.fits ...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [Done]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/lcxv13050_x1dsum.fits to data/mastDownload/HST/lcxv13050/lcxv13050_x1dsum.fits ...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [Done]
</pre></div>
</div>
</div>
</div>
<p><strong>If you’re confident Reading-in and Plotting data in Python, now go ahead to <span class="xref myst">Section 3</span></strong></p>
<p><a id="readInV"></a></p>
</section>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="reading-in-the-data">
<h1>1. Reading in the data<a class="headerlink" href="#reading-in-the-data" title="Permalink to this heading">#</a></h1>
<p>Calibrated COS 1-dimensional spectra are stored in FITS files with the suffix <code class="docutils literal notranslate"><span class="pre">x1d</span></code> or <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> (<em>1-dimensional here means that the cross-dispersion axis has been collapsed</em>).
<code class="docutils literal notranslate"><span class="pre">x1d</span></code> files contain a spectrum processed from a single exposure, while <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> files contain summed data from multiple exposures at different fixed pattern noise position settings (FP-POS). You may also encounter <code class="docutils literal notranslate"><span class="pre">x1dsumN</span></code> files, i.e. <code class="docutils literal notranslate"><span class="pre">x1dsum1</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum2</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum3</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum4</span></code>, which are intermediate files containing data from multiple exposures at the same FP-POS. All of these files are sub-types of <code class="docutils literal notranslate"><span class="pre">x1d</span></code> files and thus share the same basic data structure. What you learn with one filetype is transferable to working with another.
We will be working with <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> files in this tutorial; however, please note that we often use the shorthand “<code class="docutils literal notranslate"><span class="pre">x1d</span></code>” as part of a variable name when programming, regardless of whether it is a summed file or not.</p>
<p>The calibrated spectrum data has been downloaded onto our local machine as: <strong>&lt;<code class="docutils literal notranslate"><span class="pre">current-working-directory</span></code>&gt;</strong><code class="docutils literal notranslate"><span class="pre">/data/mastDownload/HST/</span></code><strong>&lt;<code class="docutils literal notranslate"><span class="pre">Obs_id</span></code>&gt;</strong><code class="docutils literal notranslate"><span class="pre">/</span></code><strong>&lt;<code class="docutils literal notranslate"><span class="pre">Obs_id</span></code>&gt;</strong><code class="docutils literal notranslate"><span class="pre">_x1dsum.fits</span></code>,
where the NUV and FUV Data are contained in the <code class="docutils literal notranslate"><span class="pre">obs_ids</span></code>:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Spectral Region</p></th>
<th class="head"><p>Obs_id</p></th>
<th class="head"><p>Object Name</p></th>
<th class="head"><p>Object Type</p></th>
<th class="head"><p>filepath</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>FUV</p></td>
<td><p>LCXV13050</p></td>
<td><p>QSO</p></td>
<td><p>3C48</p></td>
<td><p>./data/mastDownload/HST/lcxv13050</p></td>
</tr>
<tr class="row-odd"><td><p>NUV</p></td>
<td><p>LBBD01020</p></td>
<td><p>Star</p></td>
<td><p>WD1057 + 719</p></td>
<td><p>./data/mastDownload/HST/lbbd01020</p></td>
</tr>
</tbody>
</table>
</div>
<p><a id="investV"></a></p>
<section id="investigating-the-data-basics">
<h2>1.1. Investigating the Data - <em>Basics</em><a class="headerlink" href="#investigating-the-data-basics" title="Permalink to this heading">#</a></h2>
<p>We want to learn the basics about this file, then read in the data.
Reading the file’s primary FITS header gives us information about its data, as shown in the below cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure these filepath variables point to your new FUV data</span>
<span class="c1"># We&#39;ll often refer to the x1dsum file with the prefix x1d</span>
<span class="c1"># However, the files are x1dsum files, not the related x1d files</span>
<span class="n">fuv_x1d_filepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span>
                    <span class="s1">&#39;./data/mastDownload/HST/lcxv13050&#39;</span>
                    <span class="o">+</span> <span class="s1">&#39;/lcxv13050_x1dsum.fits&#39;</span><span class="p">)</span>

<span class="c1"># The association file; the COS calibration pipeline (CalCOS)</span>
<span class="c1"># uses this file to process spectra</span>
<span class="n">fuv_asn_filepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span>
                    <span class="s1">&#39;./data/mastDownload/HST/lcxv13050&#39;</span>
                    <span class="o">+</span> <span class="s1">&#39;/lcxv13050_asn.fits&#39;</span><span class="p">)</span>

<span class="c1"># Make sure these point to your new NUV data</span>
<span class="n">nuv_x1d_filepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span>
                    <span class="s1">&#39;./data/mastDownload/HST/lbbd01020&#39;</span>
                    <span class="o">+</span> <span class="s1">&#39;/lbbd01020_x1dsum.fits&#39;</span><span class="p">)</span>
<span class="c1"># This is the NUV association file</span>
<span class="n">nuv_asn_filepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span>
                    <span class="s1">&#39;./data/mastDownload/HST/lbbd01020&#39;</span>
                    <span class="o">+</span> <span class="s1">&#39;/lbbd01020_asn.fits&#39;</span><span class="p">)</span>

<span class="c1"># Here we get the FITS headers for our NUV/FUV x1d &amp; asn files</span>
<span class="n">nuv_x1d_header</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span>
<span class="n">nuv_asn_header</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">nuv_asn_filepath</span><span class="p">)</span>

<span class="n">fuv_x1d_header</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span>
<span class="n">fuv_asn_header</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">fuv_asn_filepath</span><span class="p">)</span>

<span class="c1"># This is the primary, 0th header of the calibrated FUV spectrum;</span>
<span class="n">fuv_x1d_header</span><span class="p">[:</span><span class="mi">18</span><span class="p">],</span> <span class="s2">&quot;...&quot;</span><span class="p">,</span> <span class="n">fuv_x1d_header</span><span class="p">[</span><span class="mi">45</span><span class="p">:</span><span class="mi">53</span><span class="p">]</span>
<span class="c1"># This is *not* the whole header!</span>
<span class="c1"># The [:18] and [45:53] tell Python</span>
<span class="c1"># to print the first 18 lines and lines 45-53</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(SIMPLE  =                    T / conforms to FITS standard                      
 BITPIX  =                    8 / array data type                                
 NAXIS   =                    0 / number of array dimensions                     
 EXTEND  =                    T                                                  
 NEXTEND =                    1 / Number of standard extensions                  
 DATE    = &#39;2024-09-10&#39;         / date this file was written (yyyy-mm-dd)        
 FILENAME= &#39;lcxv13050_x1dsum.fits&#39; / name of file                                
 FILETYPE= &#39;SCI      &#39;          / type of data found in data file                
                                                                                 
 TELESCOP= &#39;HST&#39;                / telescope used to acquire data                 
 INSTRUME= &#39;COS   &#39;             / identifier for instrument used to acquire data 
 EQUINOX =               2000.0 / equinox of celestial coord. system             
                                                                                 
               / DATA DESCRIPTION KEYWORDS                                       
                                                                                 
 ROOTNAME= &#39;lcxv13flq                         &#39; / rootname of the observation set
 IMAGETYP= &#39;TIME-TAG          &#39; / type of exposure identifier                    
 PRIMESI = &#39;COS   &#39;             / instrument designated as prime                 ,
 &#39;...&#39;,
               / DIAGNOSTIC KEYWORDS                                             
                                                                                 
 PROCTIME=   6.056377991898E+04 / pipeline processing time (MJD)                 
 OPUS_VER= &#39;HSTDP 2024_2                &#39; / data processing software system versi
 AWSYSVER= &#39;v0.4.44 &#39;           / cloud infrastructure package version           
 AWSDPVER= &#39;v0.2.26 &#39;           / cloud docker image version                     
 CSYS_VER= &#39;caldp_20240813&#39;     / calibration software system version id         
 CAL_VER = &#39;3.5.1   &#39;           / CALCOS code version                            )
</pre></div>
</div>
</div>
</div>
<p>For instance, we notice that the FUV data was taken in <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-5-cos-data-analysis/5-4-working-with-time-tag-data">TIME-TAG mode</a> and calibrated with the COS calibration pipeline, <code class="docutils literal notranslate"><span class="pre">CalCOS</span></code> version <code class="docutils literal notranslate"><span class="pre">3.4.3</span></code> (at the time of writing this sentence - it may have been reprocessed by the time you read this).</p>
<p>However, some metadata information, such as the time of observation and calculated exposure time, can be found in the <strong>1-th header</strong> rather than the 0th. We will read and print this below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Opening the FUV X1D file&#39;s header and extracting information</span>
<span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">hdu</span><span class="p">:</span>
    <span class="c1"># Opening the 1-th header</span>
    <span class="n">fuv_x1d_header1</span> <span class="o">=</span> <span class="n">hdu</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">header</span>
    <span class="c1"># Getting the date of the observation</span>
    <span class="n">fuv_date</span> <span class="o">=</span> <span class="n">fuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;DATE-OBS&#39;</span><span class="p">]</span>
    <span class="c1"># Getting the start time of the observation</span>
    <span class="n">fuv_time</span> <span class="o">=</span> <span class="n">fuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;TIME-OBS&#39;</span><span class="p">]</span>
    <span class="c1"># Getting the exposure time of the observation</span>
    <span class="n">fuv_exptime</span> <span class="o">=</span> <span class="n">fuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;EXPTIME&#39;</span><span class="p">]</span>

<span class="c1"># Note: You can also access the 1-th extension header using:</span>
<span class="c1">#    &#39;fits.getheader(fuv_x1d_filepath, ext=1)&#39;</span>

<span class="c1"># Print statement to user with newly-extracted FUV header information</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;This FUV data was taken on </span><span class="si">{</span><span class="n">fuv_date</span><span class="si">}</span><span class="s2"> starting at </span><span class="si">{</span><span class="n">fuv_time</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="sa">f</span><span class="s2">&quot; with a net exposure time of </span><span class="si">{</span><span class="n">fuv_exptime</span><span class="si">}</span><span class="s2"> seconds.&quot;</span><span class="p">)</span>

<span class="c1"># Now doing the same extraction, but with the NUV X1D file</span>
<span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">hdu</span><span class="p">:</span>
    <span class="n">nuv_x1d_header1</span> <span class="o">=</span> <span class="n">hdu</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">header</span>
    <span class="n">nuv_date</span> <span class="o">=</span> <span class="n">nuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;DATE-OBS&#39;</span><span class="p">]</span>
    <span class="n">nuv_time</span> <span class="o">=</span> <span class="n">nuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;TIME-OBS&#39;</span><span class="p">]</span>
    <span class="n">nuv_exptime</span> <span class="o">=</span> <span class="n">nuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;EXPTIME&#39;</span><span class="p">]</span>

<span class="c1"># Print statement to the user with newly-extracted NUV header info</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;This NUV data was taken on </span><span class="si">{</span><span class="n">nuv_date</span><span class="si">}</span><span class="s2"> starting at </span><span class="si">{</span><span class="n">nuv_time</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="sa">f</span><span class="s2">&quot; with a net exposure time of </span><span class="si">{</span><span class="n">nuv_exptime</span><span class="si">}</span><span class="s2"> seconds.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This FUV data was taken on 2016-06-13 starting at 23:56:29 with a net exposure time of 6532.512 seconds.
This NUV data was taken on 2009-08-14 starting at 06:03:56 with a net exposure time of 999.136 seconds.
</pre></div>
</div>
</div>
</div>
<p><a id=readmainV></a></p>
</section>
<section id="reading-in-the-x1d-x1dsum-main-data">
<h2>1.2. Reading in the <code class="docutils literal notranslate"><span class="pre">x1d</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> Main Data<a class="headerlink" href="#reading-in-the-x1d-x1dsum-main-data" title="Permalink to this heading">#</a></h2>
<p>The data are in the 1-th header-data unit (HDU) of our FITS file. There are several ways to read in our data from the our table-formatted FITS file. We’ll demonstrate three common methods below, focusing in on the <a class="reference external" href="https://docs.astropy.org/en/stable/io/unified.html#getting-started-with-table-i-o"><code class="docutils literal notranslate"><span class="pre">astropy.table</span></code></a> method.
We will then display all the fields contained in this data table using the <code class="docutils literal notranslate"><span class="pre">.colnames</span></code> method. You can ignore the warnings about multiple slashes in the units that come up while reading in the data. The proper units are displayed in LaTex as:</p>
<ul class="simple">
<li><p>‘erg /s /cm**2 /angstrom’  ==&gt; $<span class="math notranslate nohighlight">\(\ \ erg\ s^{-1}\ cm^{-2}\ \mathring{A}^{-1}\)</span>$</p></li>
<li><p>‘count /s /pixel’          ==&gt; $<span class="math notranslate nohighlight">\(\ \ counts\ s^{-1}\ pixel^{-1}\)</span>$</p></li>
</ul>
<p>The columns of these tables include some scalar values which <em>describe</em> the data (i.e. <code class="docutils literal notranslate"><span class="pre">EXPTIME</span></code>), while the columns containing actual data hold it in arrays of equal length (i.e. <code class="docutils literal notranslate"><span class="pre">WAVELENGTH</span></code>, <code class="docutils literal notranslate"><span class="pre">FLUX</span></code>, etc., where that length = <code class="docutils literal notranslate"><span class="pre">NELEM</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Method number 1:</span>
<span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">hdulist</span><span class="p">:</span>
    <span class="n">fuv_x1d_data</span> <span class="o">=</span> <span class="n">hdulist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Method 1 gets data as type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Method number 2:</span>
<span class="n">fuv_x1d_data</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getdata</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Method 2 gets data as type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Method number 3:</span>
<span class="n">fuv_x1d_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Method 3 gets data as type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># We can easily explore this astropy table</span>
<span class="n">columns</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="o">.</span><span class="n">colnames</span>
<span class="c1"># Print basic info about the table&#39;s columns</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Table of FUV data with columns:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Display a representation of the data table:</span>
<span class="n">fuv_x1d_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Method 1 gets data as type: &lt;class &#39;astropy.io.fits.fitsrec.FITS_rec&#39;&gt;
Method 2 gets data as type: &lt;class &#39;astropy.io.fits.fitsrec.FITS_rec&#39;&gt;
Method 3 gets data as type: &lt;class &#39;astropy.table.table.Table&#39;&gt;

Table of FUV data with columns:
 [&#39;SEGMENT&#39;, &#39;EXPTIME&#39;, &#39;NELEM&#39;, &#39;WAVELENGTH&#39;, &#39;FLUX&#39;, &#39;ERROR&#39;, &#39;ERROR_LOWER&#39;, &#39;GROSS&#39;, &#39;GCOUNTS&#39;, &#39;VARIANCE_FLAT&#39;, &#39;VARIANCE_COUNTS&#39;, &#39;VARIANCE_BKG&#39;, &#39;NET&#39;, &#39;BACKGROUND&#39;, &#39;DQ&#39;, &#39;DQ_WGT&#39;] 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
</pre></div>
</div>
<div class="output text_html"><div><i>Table length=2</i>
<table id="table140470740924240" class="table-striped table-bordered table-condensed">
<thead><tr><th>SEGMENT</th><th>EXPTIME</th><th>NELEM</th><th>WAVELENGTH</th><th>FLUX</th><th>ERROR</th><th>ERROR_LOWER</th><th>GROSS</th><th>GCOUNTS</th><th>VARIANCE_FLAT</th><th>VARIANCE_COUNTS</th><th>VARIANCE_BKG</th><th>NET</th><th>BACKGROUND</th><th>DQ</th><th>DQ_WGT</th></tr></thead>
<thead><tr><th></th><th>s</th><th></th><th>Angstrom</th><th>erg / (Angstrom s cm2)</th><th>erg / (Angstrom s cm2)</th><th>erg / (Angstrom s cm2)</th><th>ct / s</th><th>ct</th><th>ct</th><th>ct</th><th>ct</th><th>ct / s</th><th>ct / s</th><th></th><th></th></tr></thead>
<thead><tr><th>bytes4</th><th>float64</th><th>int32</th><th>float64[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>float32[16384]</th><th>int16[16384]</th><th>float32[16384]</th></tr></thead>
<tr><td>FUVA</td><td>6532.512</td><td>16384</td><td>1610.240833382022 .. 1810.9448636791044</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>128 .. 128</td><td>0.0 .. 0.0</td></tr>
<tr><td>FUVB</td><td>6532.512</td><td>16384</td><td>1421.9464241857759 .. 1622.5957869332851</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>128 .. 128</td><td>0.0 .. 0.0</td></tr>
</table></div></div></div>
</div>
<p><strong>In the case of the FUV data, we see an <code class="docutils literal notranslate"><span class="pre">astropy</span></code> style table of 2 rows</strong> which are labeled FUVA and FUVB. These rows contain data from the 2 segments of FUV Detector (see Figure 1.1).</p>
<section id="fig-1-1-from-cos-dhb-fig-1-6">
<h3>Fig. 1.1 from <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-1-cos-overview/1-2-cos-physical-configuration#id-1.2COSPhysicalConfiguration-Figure1.6">COS DHB Fig. 1.6</a><a class="headerlink" href="#fig-1-1-from-cos-dhb-fig-1-6" title="Permalink to this heading">#</a></h3>
<center>
    <img src=figures/cosdhb_fig1p6.jpg width ="900" title="Layout of the COS FUV detector">
    <figcaption>Layout of the COS FUV detector. Note that FUVB corresponds to shorter wavelengths than FUVA.</figcaption>
</center>
<p><strong>In the case of the NUV data, we see a similar <code class="docutils literal notranslate"><span class="pre">astropy</span></code> style table of 3 rows</strong> (labeled NUVA, NUVB, and NUVC). These rows contain data from the 3 stripes of the NUV spectrum (see Figure 1.2).</p>
</section>
<section id="fig-1-2-from-cos-dhb-fig-1-10">
<h3>Fig. 1.2 from <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-1-cos-overview/1-2-cos-physical-configuration#id-1.2COSPhysicalConfiguration-Figure1.10">COS DHB Fig. 1.10</a><a class="headerlink" href="#fig-1-2-from-cos-dhb-fig-1-10" title="Permalink to this heading">#</a></h3>
<center>
    <img src=figures/ch1_cos_overview3.10.jpg width ="900" title="An example COS NUV spectrum">
    <figcaption>An example COS NUV spectrum. The spectrum itself, taken with the Primary Science Aperture, is in the lower three stripes labeled 'PSA'. The upper stripes, labeled 'WCA' are for wavelength calibration.</figcaption>
</center>
<p>An important thing to note about this <em>NUV</em> data in particular is that with the grating used here (G230L), stripe C is actually a 2nd order spectrum with a higher dispersion and ~5% contamination from the 1st order spectrum. See the <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-1-cos-overview/1-1-instrument-capabilities-and-design#id-1.1InstrumentCapabilitiesandDesign-NUVSpectroscopyNUVSpectroscopy">COS Data Handbook</a> for more information.</p>
<p><a id=assocV></a></p>
</section>
</section>
<section id="the-association-asn-file">
<h2>1.3. The Association (<code class="docutils literal notranslate"><span class="pre">asn</span></code>) file<a class="headerlink" href="#the-association-asn-file" title="Permalink to this heading">#</a></h2>
<p>It’s also likely we will want to see what observations went into making this calibrated spectrum. This information is contained in the Association (<code class="docutils literal notranslate"><span class="pre">asn</span></code>) file, under the <code class="docutils literal notranslate"><span class="pre">MEMNAME</span></code> column, which can be seen after running the next cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Printing the summary information about the FITS file</span>
<span class="nb">print</span><span class="p">(</span><span class="n">fits</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">fuv_asn_filepath</span><span class="p">),</span> <span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">----</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># Getting the asn file data in an astropy table</span>
<span class="n">fuv_asn_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fuv_asn_filepath</span><span class="p">)</span>
<span class="c1"># Printing said table</span>
<span class="nb">print</span><span class="p">(</span><span class="n">fuv_asn_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Filename: data/mastDownload/HST/lcxv13050/lcxv13050_asn.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU      43   ()      
  1  ASN           1 BinTableHDU     25   5R x 3C   [14A, 14A, L]   
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>None 

----

   MEMNAME        MEMTYPE     MEMPRSNT
-------------- -------------- --------
     LCXV13FLQ         EXP-FP        1
     LCXV13FXQ         EXP-FP        1
     LCXV13G4Q         EXP-FP        1
     LCXV13GXQ         EXP-FP        1
     LCXV13050        PROD-FP        1
</pre></div>
</div>
</div>
</div>
<p>We see that our data has <code class="docutils literal notranslate"><span class="pre">MEMTYPE</span></code> = <code class="docutils literal notranslate"><span class="pre">PROD-FP</span></code>, meaning it is an output science product (see COS DHB <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-2-cos-data-files/2-4-cos-data-products#:~:text=Figure%202.6%3A%20FITS%20File%20Format%20for%201%2DD%20Extracted%20Spectrum%20Table.">Table 2.6</a>.)</p>
<p>This particular association file lists four <code class="docutils literal notranslate"><span class="pre">EXP-FP</span></code> (input science exposure), with the <code class="docutils literal notranslate"><span class="pre">MEMNAME</span></code> values (Dataset IDs) <code class="docutils literal notranslate"><span class="pre">[LCXV13FLQ,</span> <span class="pre">LCXV13FXQ,</span> <span class="pre">LCXV13G4Q,</span> <span class="pre">LCXV13GXQ]</span></code>. We could look for these datasets, if we wished to inspect the exposures individually.</p>
<section id="exercise-1-1-finding-metadata-for-the-nuv">
<h3>Exercise 1.1. <em>Finding Metadata for the NUV</em><a class="headerlink" href="#exercise-1-1-finding-metadata-for-the-nuv" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Read in the NUV data, just as we did with the FUV data.</p></li>
<li><p>From the 0th header of the <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> file, determine the time (in MJDs) that the data was processed. (keyword = <code class="docutils literal notranslate"><span class="pre">PROCTIME</span></code>) and from the 1th header, determine how many wavelength calibration “flashes” were used (keyword = <code class="docutils literal notranslate"><span class="pre">NUMFLASH</span></code>)</p></li>
<li><p>From the <code class="docutils literal notranslate"><span class="pre">asn</span></code> file, determine how many input science exposures went into the NUV <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> file.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Your code here</span>
</pre></div>
</div>
</div>
</div>
<p><a id="plottingV"></a></p>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="plotting-the-data">
<h1>2. Plotting the Data<a class="headerlink" href="#plotting-the-data" title="Permalink to this heading">#</a></h1>
<p><a id = plotdataFUV></a></p>
<section id="plotting-an-fuv-spectrum">
<h2>2.1 Plotting an FUV Spectrum<a class="headerlink" href="#plotting-an-fuv-spectrum" title="Permalink to this heading">#</a></h2>
<p>In <span class="xref myst">Section 1.2</span>, we read in our <code class="docutils literal notranslate"><span class="pre">x1d</span></code> and <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> files’ data using the three different method. By working with the rows and columns of our table, we can plot our spectrum as wavelength vs flux for each segment.</p>
<p><a id = firstplotFUV></a></p>
<section id="our-first-plot">
<h3>2.1.1 Our First Plot<a class="headerlink" href="#our-first-plot" title="Permalink to this heading">#</a></h3>
<p>Our table has two rows: the first for Segment A and the second for Segment B. We’ll demonstrate how to plot the data for Segment A first. We can specify the row we’re interested in using the index of the row (an integer); our columns can be accessed in a similar manner, using the column name of interest (which will be a string).</p>
<p>Below, we use the form <code class="docutils literal notranslate"><span class="pre">our_table[ROW][&quot;COLUMN&quot;]</span></code> to access our data but you could just as easily use <code class="docutils literal notranslate"><span class="pre">our_table[&quot;COLUMN&quot;][ROW]</span></code> instead. <em>Note that there is a difference in speed between the two. <code class="docutils literal notranslate"><span class="pre">astropy</span></code> recommends using the second form since the code will not create an intermediate <code class="docutils literal notranslate"><span class="pre">Row</span></code> object.</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access FUVA data</span>
<span class="n">wvln</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># Set up the plot as a single box with a figure size of 10x4 inches,</span>
<span class="c1"># and with a dpi of 100 (dpi = px per inch).</span>
<span class="c1"># The first two params here correspond to the num of rows/cols,</span>
<span class="c1"># they&#39;re both 1 here since we&#39;re plotting a single</span>
<span class="c1"># plot but this is going to be useful when we work</span>
<span class="c1"># with multiple subplots.</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                       <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1">##############</span>
<span class="c1"># Here we plot wavelength vs flux:</span>
<span class="c1"># The x and y values for our plot, wavelength and flux here</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
        <span class="c1"># This sets the thickness for the plot line</span>
        <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
        <span class="c1"># Here we&#39;re setting the color of the line</span>
        <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="c1">##############</span>
<span class="c1"># We&#39;re adding the labels and title to the plot now:</span>
<span class="c1"># Setting the x-axis label</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
<span class="c1"># Setting the y-axis label</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

<span class="c1"># Adding the title to the plot and putting</span>
<span class="c1"># the FUV segment from the FITS header in the title too</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Fig 2.1</span><span class="se">\n</span><span class="s1">Simple COS Segment &#39;</span> <span class="o">+</span> <span class="n">segment</span> <span class="o">+</span> <span class="s1">&#39; Spectrum&#39;</span><span class="p">,</span>
             <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="c1"># Now we can save the figure to our plots directory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s2">&quot;Fig2.1.png&quot;</span><span class="p">))</span>

<span class="c1"># We can show the figure below once the cell is run</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/b36b01a10f05de69790b09a5a95b9bbffc7247fd72e5f2dc58a12523c64d76f1.png" src="../../../_images/b36b01a10f05de69790b09a5a95b9bbffc7247fd72e5f2dc58a12523c64d76f1.png" />
</div>
</div>
</section>
<section id="exercise-2-1-a-clearer-plot">
<h3>Exercise 2.1: A clearer plot<a class="headerlink" href="#exercise-2-1-a-clearer-plot" title="Permalink to this heading">#</a></h3>
<p>Plot the data from Segment FUVB, similar to the Segment FUVA plot above.</p>
<p>Note what can you simply copy over, and what you have to make sure to change.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Your code here</span>
</pre></div>
</div>
</div>
</div>
<p><a id = plotcompV></a></p>
</section>
<section id="a-complex-look-at-the-entire-fuv">
<h3>2.1.2. A Complex Look at the Entire FUV<a class="headerlink" href="#a-complex-look-at-the-entire-fuv" title="Permalink to this heading">#</a></h3>
<p><strong>Now that we have an idea for how <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> works, let’s make a more complicated graph showing both FUV segments - <em>independently and together</em>.</strong></p>
<p>One of the most important steps to creating a plot is planning out how it will look and convey its information. We’ll begin by planning this out below:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Panel</p></th>
<th class="head"><p>Contents</p></th>
<th class="head"><p>Information Conveyed</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><em>top</em></p></td>
<td><p>Entire FUV Spectrum as a <strong>simple plot</strong></p></td>
<td><p>Overview of the entire spectrum we have, coarse look without much detail</p></td>
<td><p>Color by segment</p></td>
</tr>
<tr class="row-odd"><td><p><em>middle</em></p></td>
<td><p>Shorter Wavelength FUVB Spectrum as an <strong>errorbar plot</strong></p></td>
<td><p>Closer look at the shorter wavelength spectrum with an idea of error</p></td>
<td><p>Color errorbar by segment, central line in black</p></td>
</tr>
<tr class="row-even"><td><p><em>bottom</em></p></td>
<td><p>Longer Wavelength FUVA Spectrum as an <strong>errorbar plot</strong></p></td>
<td><p>Closer look at the longer wavelength spectrum with an idea of error</p></td>
<td><p>Color errorbar by segment, central line in black</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ax0, ax1, ax2 are our 3 vertically-aligned panels,</span>
<span class="c1"># which are top, middle, and bottom.</span>
<span class="c1"># &quot;fig&quot; is the entire figure, which consists of the three subplots.</span>
<span class="c1"># Here we have 3 rows, 1 column, figure size of 10x10 inches</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="c1"># Repeats for i=[0,1] to apply to each segment&#39;s data at a time</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="c1"># Selects all useful data for the chosen segment</span>
    <span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">,</span> <span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">fluxErr</span><span class="p">,</span> <span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>

    <span class="c1"># This section applies the top panel&#39;s plotting and formatting:</span>
    <span class="c1"># Plotting wavelength vs flux:</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
             <span class="c1"># Labeling our plot with its segment (used in legend)</span>
             <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
             <span class="c1"># Getting the segment color for FUV/NUV A/B</span>
             <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

    <span class="c1"># Formatting the plot in the next few cells:</span>
    <span class="c1"># Getting a legend for our plot, located in the upper left</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
    <span class="c1"># Setting the title of the plot. &quot;size&quot; here is our fontsize</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Fig 2.2</span><span class="se">\n</span><span class="s2">FUV Spectra with G160M Grating&quot;</span><span class="p">,</span>
                  <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
    <span class="c1"># Setting the limit for the x axis; ax.set_xlim(START, END)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">1428</span><span class="p">,</span> <span class="mi">1805</span><span class="p">)</span>
    <span class="c1"># Setting the limits for the y axis; ax.set_ylim(START, END)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">1E-15</span><span class="p">,</span> <span class="mf">1.9E-14</span><span class="p">)</span>

    <span class="c1">######</span>
    <span class="c1"># Indented code applies only to segment FUVA data in bottom panel</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Plotting wavelength vs flux, with y axis error (flux)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                     <span class="c1"># Setting the color of our errorbars</span>
                     <span class="n">ecolor</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">1610</span><span class="p">,</span> <span class="mi">1810</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">3E-15</span><span class="p">,</span> <span class="mf">1.9E-14</span><span class="p">)</span>
        <span class="c1"># Putting our legend in the upper left portion of the plot</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
                       <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

    <span class="c1"># Plotting with errorbars for the FUVB data, similar to FUVA plot</span>
    <span class="c1"># This indented code applies only to FUVB data in middle panel</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                     <span class="n">ecolor</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">1428</span><span class="p">,</span> <span class="mi">1615</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">1E-15</span><span class="p">,</span> <span class="mf">1.25E-14</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
                       <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="c1"># Clipping the whitespace of the figure</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="c1"># Saving our figure to the plots directory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig2.2.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="c1"># Showing our figure below once we run the cell</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/90720ae9e4a596af6d63051558e6bd07f5846321ec4274462ad5f565838bf82e.png" src="../../../_images/90720ae9e4a596af6d63051558e6bd07f5846321ec4274462ad5f565838bf82e.png" />
</div>
</div>
<p><a id = plotcloserV></a></p>
</section>
<section id="looking-closer-at-parts-of-the-fuv-spectrum">
<h3>2.1.3. Looking Closer at Parts of the FUV Spectrum<a class="headerlink" href="#looking-closer-at-parts-of-the-fuv-spectrum" title="Permalink to this heading">#</a></h3>
<p><strong>It can be very difficult to get any insights on small-scale details in the above plots because data is too dense to parse at once. Below, we’ll show examples of:</strong></p>
<ul class="simple">
<li><p>Plotting a small region around an absorption line feature.</p></li>
<li><p>Plotting an entire segment’s spectrum in segments of wavelength space.</p></li>
</ul>
<p><strong>Let’s begin with showing a region around a line, in this case, we see a sharp line around 1670 Å.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># By-eye center of the line for plotting</span>
<span class="n">line1670</span> <span class="o">=</span> <span class="mf">1670.67</span>

<span class="c1"># Selects all useful data for the chosen segment</span>
<span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">,</span> <span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">fluxErr</span><span class="p">,</span> <span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># How many Angstroms in each direction around line 1670 do we want to look at?</span>
<span class="n">wvln_extent</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># Mask the data to within +/- wvln_extent Angstrom of the line</span>
<span class="n">lineRegion_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&gt;</span> <span class="n">line1670</span> <span class="o">-</span> <span class="n">wvln_extent</span><span class="p">)</span>\
                  <span class="o">&amp;</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&lt;</span> <span class="n">line1670</span> <span class="o">+</span> <span class="n">wvln_extent</span><span class="p">)</span>

<span class="c1"># Applies the created mask so we get data from our desired wavelength range</span>
<span class="n">wvln_region</span> <span class="o">=</span> <span class="n">wvln</span><span class="p">[</span><span class="n">lineRegion_mask</span><span class="p">]</span>
<span class="n">flux_region</span> <span class="o">=</span> <span class="n">flux</span><span class="p">[</span><span class="n">lineRegion_mask</span><span class="p">]</span>
<span class="n">fluxErr_region</span> <span class="o">=</span> <span class="n">fluxErr</span><span class="p">[</span><span class="n">lineRegion_mask</span><span class="p">]</span>

<span class="c1"># Setting the size of the plot, in inches</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Creating a verticle line on the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">line1670</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">2.5</span><span class="p">,</span>
            <span class="c1"># Setting the style of the line</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$Approximate\ line\ center$&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln_region</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux_region</span><span class="p">,</span>
             <span class="c1"># Plotting an error bar</span>
             <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr_region</span><span class="p">,</span>
             <span class="c1"># Labeling errorbar for legend</span>
             <span class="n">label</span><span class="o">=</span><span class="s1">&#39;COS FUVA&#39;</span><span class="p">,</span>
             <span class="c1"># Setting the marker for the errorbar</span>
             <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Fig 2.3, Zoom in on the line at 1670 $\AA$&#39;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="c1"># Saving the figure to plots directory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig2.3.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/d82015671fac2a5fd2193a8e8a67a6d833e90b6c1fa4765fe961873294340081.png" src="../../../_images/d82015671fac2a5fd2193a8e8a67a6d833e90b6c1fa4765fe961873294340081.png" />
</div>
</div>
<p>The source we’re looking at - the quasar 3C48 has a substantial redshift: $<span class="math notranslate nohighlight">\(z \approx 0.37\)</span>$</p>
<p>Source: <a class="reference external" href="http://simbad.u-strasbg.fr/simbad/sim-basic?Ident=3c48&amp;submit=SIMBAD+search">Simbad search for 3C48</a></p>
<p>With that redshift, we can determine the rest-frame wavelength.</p>
<div class="math notranslate nohighlight">
\[z = \dfrac{\lambda_{obs}-\lambda_{rest}}{\lambda_{rest}} \ \ \rightarrow\ \ \ \lambda_{rest} = \dfrac{\lambda_{obs}}{1+z}\]</div>
<p>At that redshift, we see this line to be <span class="math notranslate nohighlight">\(\approx 1219\)</span> Å. This is close to, and may correspond to, the Lyman-<span class="math notranslate nohighlight">\(\alpha\)</span> transition.</p>
<p><font size="4"><b>Now we will create a much more complex plot,</b></font> <font size="3">allowing us to visualize an entire segment’s spectrum in fine detail</font></p>
<p>We’ll split the spectrum into “segments” of ~10 Angstroms, and plot these segments in a vertical series. These plots may take a minute to create. If you want them to run quicker, hit the <code class="docutils literal notranslate"><span class="pre">Interrupt</span> <span class="pre">the</span> <span class="pre">Kernel</span></code> square-shaped button to stop the current cell from running (you will get an error message), and run this cell with a smaller number of rows (<code class="docutils literal notranslate"><span class="pre">nRows</span></code>).
<a id="fig2p4V"></a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply the following to each segment&#39;s data at a time</span>
<span class="k">for</span> <span class="n">segment_row</span> <span class="ow">in</span> <span class="n">fuv_x1d_data</span><span class="p">:</span>
    <span class="c1"># Selects all immediately useful data for the chosen segment</span>
    <span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">segment_row</span><span class="p">[</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">,</span> <span class="s1">&#39;FLUX&#39;</span><span class="p">]</span>
    <span class="n">fluxErr</span><span class="p">,</span> <span class="n">segment</span> <span class="o">=</span> <span class="n">segment_row</span><span class="p">[</span><span class="s1">&#39;ERROR&#39;</span><span class="p">,</span> <span class="s1">&#39;SEGMENT&#39;</span><span class="p">]</span>

    <span class="c1"># Getting the min and max wavelength and flux values for segment</span>
    <span class="n">minx</span><span class="p">,</span> <span class="n">maxx</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">wvln</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">wvln</span><span class="p">)</span>
    <span class="n">miny</span><span class="p">,</span> <span class="n">maxy</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">flux</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">flux</span><span class="p">)</span>
    <span class="c1"># Getting the wavelength range for segment</span>
    <span class="n">rangex</span> <span class="o">=</span> <span class="n">maxx</span> <span class="o">-</span> <span class="n">minx</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>

    <span class="c1"># How many segments we wish to split the spectrum into</span>
    <span class="n">nRows</span> <span class="o">=</span> <span class="mi">15</span>

    <span class="c1"># Going through each segment and plotting its wavelength range</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nRows</span><span class="p">):</span>
        <span class="c1"># Getting the min and max wavelength range for this segment</span>
        <span class="n">min_</span> <span class="o">=</span> <span class="n">minx</span> <span class="o">+</span> <span class="n">i</span> <span class="o">*</span> <span class="n">rangex</span> <span class="o">/</span> <span class="n">nRows</span>
        <span class="n">max_</span> <span class="o">=</span> <span class="n">minx</span> <span class="o">+</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">rangex</span> <span class="o">/</span> <span class="n">nRows</span>

        <span class="c1"># Creating a subplot for this segment</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">nRows</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Set the title, xlabel, and ylabel, independent of # of rows:</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;Fig. 2.4</span><span class="si">{</span><span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1"> </span><span class="se">\n</span><span class="s1">Segment </span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s1"> &#39;</span>
                <span class="s1">&#39;Spectrum split into segments&#39;</span><span class="p">,</span>
                <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">nRows</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
                          <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">nRows</span> <span class="o">/</span> <span class="mi">2</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
                          <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

        <span class="c1"># Creating the plot itself:</span>
        <span class="c1"># Plotting wavelength vs flux, with flux error</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span> <span class="n">fluxErr</span><span class="p">,</span>
                    <span class="n">c</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">rainbow</span><span class="p">((</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">nRows</span><span class="p">),</span>
                    <span class="c1"># Setting the transparency</span>
                    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span>
                    <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
                    <span class="n">markerfacecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                    <span class="c1"># Setting the size of the marker</span>
                    <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                    <span class="n">mew</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_</span><span class="p">,</span> <span class="n">max_</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;Fig2.4</span><span class="si">{</span><span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span>
                    <span class="s1">&#39;_</span><span class="si">{nRows}</span><span class="s1">Rows_seg</span><span class="si">{segment}</span><span class="s1">.png&#39;</span><span class="p">),</span>
                <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="c1"># Printing this line to separate FUVA and FUVB&#39;s plots</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">----</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/2a56c8d843e68539dd874095f286a31301549fe33e28069b73eb703e056263d2.png" src="../../../_images/2a56c8d843e68539dd874095f286a31301549fe33e28069b73eb703e056263d2.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>----
</pre></div>
</div>
<img alt="../../../_images/718ceb707265dd71c810a03962d43d06f6b710c606af9a7c16405ddda0e6ea81.png" src="../../../_images/718ceb707265dd71c810a03962d43d06f6b710c606af9a7c16405ddda0e6ea81.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>----
</pre></div>
</div>
</div>
</div>
</section>
<section id="exercise-2-2-choosing-and-plotting-a-useful-data-range">
<h3>Exercise 2.2 Choosing and plotting a useful data range<a class="headerlink" href="#exercise-2-2-choosing-and-plotting-a-useful-data-range" title="Permalink to this heading">#</a></h3>
<p>That last plot was very dense with information. As practice plotting, make a scaled down version from just <code class="docutils literal notranslate"><span class="pre">1635</span> <span class="pre">-</span> <span class="pre">1675</span></code> Angstroms, and plot the data as a simple line graph.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Your code here</span>
</pre></div>
</div>
</div>
</div>
<p><a id = specvizV></a></p>
</section>
<section id="reading-and-plotting-the-data-with-specutils-optional">
<h3>2.1.4. Reading and plotting the data with <code class="docutils literal notranslate"><span class="pre">specutils</span></code> (<em>OPTIONAL</em>)<a class="headerlink" href="#reading-and-plotting-the-data-with-specutils-optional" title="Permalink to this heading">#</a></h3>
<p><font size="4.5"> <em>Note, this section is entirely optional, and simply meant to give you more options for how to plot a spectrum.</em></font></p>
<p>An alternative way to read in and work with spectral data is with the <a class="reference external" href="https://specutils.readthedocs.io/en/stable/"><code class="docutils literal notranslate"><span class="pre">specutils</span></code> package</a>, which contains <em>quite a bit of functionality</em> for working with spectra. It also can make dealing with units easier, as it generally works well with astropy units and other modules. Make sure that your package is up to date (version <span class="math notranslate nohighlight">\(\ge\)</span> 1.1). We describe installing <code class="docutils literal notranslate"><span class="pre">specutils</span></code> in our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/blob/main/notebooks/COS/Setup/Setup.ipynb">Notebook on setting up an environment</a>.</p>
<p><code class="docutils literal notranslate"><span class="pre">specutils</span></code> treats spectra as special python objects rather than lists, and reads in the entire spectrum over both segments.</p>
<p>Below is a simple example of using <code class="docutils literal notranslate"><span class="pre">specutils</span></code> to read-in, plot, and continuum-normalize our entire spectrum.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">specutils</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">specutils.fitting</span><span class="w"> </span><span class="kn">import</span> <span class="n">fit_generic_continuum</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">rcParams</span>

<span class="c1"># The default fontsize it will use does not be clear at this plot size</span>
<span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;font.size&#39;</span><span class="p">:</span> <span class="mi">20</span><span class="p">})</span>

<span class="c1"># Here we&#39;re getting the entire spectrum (FUVA &amp; FUVB)</span>
<span class="n">spec1d</span> <span class="o">=</span> <span class="n">specutils</span><span class="o">.</span><span class="n">Spectrum1D</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span>

<span class="c1"># 2 rows, 1 column of subplots, 10x8 in. figure</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="c1"># Plot the non-normalized flux</span>
<span class="c1"># Wavelength vs non-normalized flux</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">spec1d</span><span class="o">.</span><span class="n">spectral_axis</span><span class="p">,</span> <span class="n">spec1d</span><span class="o">.</span><span class="n">flux</span><span class="p">)</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Fig. 2.5</span><span class="se">\n</span><span class="s2">Producing spectra with &quot;</span>
              <span class="o">+</span> <span class="s2">&quot;specutils</span><span class="se">\n</span><span class="s2">Un-normalized flux&quot;</span><span class="p">)</span>

<span class="c1"># Continuum Normalize the flux:</span>
<span class="n">cont_norm_spec1d</span> <span class="o">=</span> <span class="n">spec1d</span> <span class="o">/</span>\
                   <span class="n">fit_generic_continuum</span><span class="p">(</span><span class="n">spec1d</span><span class="p">)(</span><span class="n">spec1d</span><span class="o">.</span><span class="n">spectral_axis</span><span class="p">)</span>

<span class="c1"># Plot the normalized flux and setting the title and y axis label</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cont_norm_spec1d</span><span class="o">.</span><span class="n">spectral_axis</span><span class="p">,</span> <span class="n">cont_norm_spec1d</span><span class="o">.</span><span class="n">flux</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Normalized flux&quot;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Unitless&quot;</span><span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig2.5.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># Not necessary for this plot, but gives us all the data separately</span>
<span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">spec1d</span><span class="o">.</span><span class="n">wavelength</span><span class="p">,</span> <span class="n">spec1d</span><span class="o">.</span><span class="n">flux</span>
<span class="n">fluxErr</span> <span class="o">=</span> <span class="n">spec1d</span><span class="o">.</span><span class="n">uncertainty</span><span class="o">.</span><span class="n">quantity</span>
<span class="c1"># Restore default fontsize</span>
<span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;font.size&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: Model is linear in parameters; consider using linear fitting methods. [astropy.modeling.fitting]
</pre></div>
</div>
<img alt="../../../_images/26ab940a319b9efd48864924680003ed4dfbbae573e3f3c68e78239e80ab71f1.png" src="../../../_images/26ab940a319b9efd48864924680003ed4dfbbae573e3f3c68e78239e80ab71f1.png" />
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">specutils</span></code> also works with the visualization <a class="reference external" href="https://github.com/spacetelescope/jdaviz">package <code class="docutils literal notranslate"><span class="pre">Jdaviz.Specviz</span></code></a>, to produce very useful interactive spectral plots within a Jupyter Notebook GUI framework. Currently, this interactive program works well within the Jupyter Notebook environment, but not within Jupyter Lab environment.</p>
<p><a id=nuvV></a></p>
</section>
</section>
<section id="plotting-an-nuv-spectrum">
<h2>2.2. Plotting an NUV Spectrum<a class="headerlink" href="#plotting-an-nuv-spectrum" title="Permalink to this heading">#</a></h2>
<p>The NUV data is formatted very similarly to the FUV data, and can be accessed as in <span class="xref myst">Section 1</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Here we read in the data</span>
<span class="c1"># We may ignore warnings about slashes in the notation</span>
<span class="n">nuv_x1d_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;count /s /pixel&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
</pre></div>
</div>
</div>
</div>
<p><a id = nuv1V></a></p>
<section id="examining-the-first-order-spectrum">
<h3>2.2.1. Examining the first-order spectrum<a class="headerlink" href="#examining-the-first-order-spectrum" title="Permalink to this heading">#</a></h3>
<p><strong>Again, let’s select the simplest data we need to plot a spectrum: WAVELENGTH, FLUX, and ERROR</strong>.</p>
<p>We will limit this first plot to only our first order spectra (we will exclude stripe C). We will plot these two spectra together (in the top panel) and then stripe-by-stripe in the lower 2 panes, similar to Figure 2.2.</p>
<ul class="simple">
<li><p>The former view gives a better sense of the continuum</p></li>
<li><p>The latter views show zooms which allow us to meaningfully view the errors and see specific emission/absorption features.</p></li>
</ul>
<p>The way we produce this plot with <code class="docutils literal notranslate"><span class="pre">Python</span></code> in the cell below is looping through both NUV first-order stripes - NUVA and NUVB - plotting both of them in the upper subplot, and plotting each of them once more in one of the lower two subplots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Build a 3row x 1column - subplot figure</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">16</span><span class="p">))</span>

<span class="c1"># Iterate through the first 2 NUV segments/stripes</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="c1"># Gathering the data</span>
    <span class="n">wvln</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>
    <span class="n">fluxErr</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;ERROR&quot;</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">]</span>

    <span class="n">ax0</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
             <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
             <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

    <span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Fig 2.6</span><span class="se">\n</span><span class="s2">First-order NUV Spectra G230L Grating&quot;</span><span class="p">,</span>
                  <span class="n">size</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>

    <span class="c1"># Plot only 0th segment/stripe in 2nd panel</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Plotting wavelength vs flux with errorbars</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="c1"># Linestyle is &quot;&quot; because only plotting fluxErr</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">2100</span><span class="p">,</span> <span class="mi">2510</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">0.5E-13</span><span class="p">,</span> <span class="mf">1.5E-13</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
                       <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

    <span class="c1"># Plot only 1th segment/stripe in 3rd panel, similar to above plot</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">3190</span><span class="p">,</span> <span class="mi">3600</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">1E-14</span><span class="p">,</span> <span class="mf">4.5E-14</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
                       <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig2.6.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/727a850c04feabba4ac35eac0d3e9fee5df645a42ad86e58666eb844be069be5.png" src="../../../_images/727a850c04feabba4ac35eac0d3e9fee5df645a42ad86e58666eb844be069be5.png" />
</div>
</div>
<p><a id = nuv2V></a></p>
</section>
<section id="examining-the-second-order-spectrum">
<h3>2.2.2. Examining the second-order spectrum<a class="headerlink" href="#examining-the-second-order-spectrum" title="Permalink to this heading">#</a></h3>
<p>On NUV stripe C data taken with the grating G230L, we have a more dispersed, second-order spectrum over a smaller segment of the NUV. Below, we plot this second-order over the first-order spectrum from stripe A.</p>
<p>Note that most NUV gratings produce first-order spectra over all three stripes. See the <a class="reference external" href="https://hst-docs.stsci.edu/cosihb/chapter-5-spectroscopy-with-cos/5-1-the-capabilities-of-cos#id-5.1TheCapabilitiesofCOS-Section5.1.35.1.3Second-OrderSensitivity">COS Instrument Handbook</a> for more information. We chose to plot this this type of spectrum to create a more complex plot and to demonstrate using different scales on the same axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Creating a figure with 2 subplots:</span>
<span class="c1"># 2 rows, 1 column, size 16x8 inches</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="c1"># We reverse this order. The 0th stripe (A) is plotted OVER the 2nd stripe (C)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]:</span>
    <span class="c1"># Gathering the data for our NUV stripe</span>
    <span class="n">wvln</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;FLUX&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">fluxErr</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;ERROR&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;SEGMENT&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>

    <span class="c1"># This is for NUVA stripe</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ax0</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="c1"># Plotting errorbars for NUVA bottom plot</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

    <span class="c1"># This is for NUVC stripe data, same style parameters as the NUVA</span>
    <span class="c1"># segment but different values for those parameters</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">ax0</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">],</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">wvln</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">flux</span><span class="p">,</span>
                     <span class="n">yerr</span><span class="o">=</span><span class="n">fluxErr</span><span class="p">,</span>
                     <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                     <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
                     <span class="n">c</span><span class="o">=</span><span class="n">segment_colors</span><span class="p">[</span><span class="n">segment</span><span class="p">],</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>

<span class="c1"># Setting the top plot&#39;s x and y axis limits</span>
<span class="c1"># These are wavelength and flux respectively</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">2100</span><span class="p">,</span> <span class="mi">2510</span><span class="p">)</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">0.48E-13</span><span class="p">,</span> <span class="mf">1.99E-13</span><span class="p">)</span>

<span class="c1"># Creating a title for the overall plot</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Fig 2.7</span><span class="se">\n</span><span class="s1">Overlay of First and &#39;</span> <span class="o">+</span>
              <span class="s1">&#39;Second-order Spectrum with G230L Grating&#39;</span><span class="p">,</span>
              <span class="n">size</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>

<span class="c1"># Setting the x and y axis limits for the bottom plot</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">2200</span><span class="p">,</span> <span class="mi">2210</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">0.6E-13</span><span class="p">,</span> <span class="mf">1.7E-13</span><span class="p">)</span>

<span class="c1"># Setting the x axis label for the plots</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
               <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="c1"># Creating text for the figure that is the y axis text of both plots</span>
<span class="c1"># Coordinates for text box, actual text</span>
<span class="n">fig</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="o">-</span><span class="mf">0.015</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span>
         <span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
         <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
         <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
         <span class="c1"># Specifying rotation</span>
         <span class="n">rotation</span><span class="o">=</span><span class="s1">&#39;vertical&#39;</span><span class="p">)</span>

<span class="c1"># Adding 2 vertical lines to show wavelength bounds on the subplot</span>
<span class="c1"># Leftmost line at 2200 angstroms</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">2200</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;Lower panel $\lambda$ bounds&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>

<span class="c1"># Rightmost line at 2210 angstroms</span>
<span class="n">ax0</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">2210</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>

<span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig2.7.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/73922f1d72a89d6d2afdd8a41316dad2ce01b7b1afc929e973033a5a07bbad6d.png" src="../../../_images/73922f1d72a89d6d2afdd8a41316dad2ce01b7b1afc929e973033a5a07bbad6d.png" />
</div>
</div>
<p><strong>Clearly, our errorbars on the second-order spectrum (NUVC) are <em>much larger</em>.</strong></p>
<p>However, if we need a very high dispersion - for instance, to split closely-spaced lines - the lower panel shows a potential advantage of stripe C. Its higher spectral sampling rate can allow for finer distinctions in wavelength, <em>if</em> an acceptable SNR can be reached.</p>
<p><a id = AssessingV></a></p>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="preliminary-review-of-the-data">
<h1>3. Preliminary Review of the Data<a class="headerlink" href="#preliminary-review-of-the-data" title="Permalink to this heading">#</a></h1>
<p><font size="4"><b>There is no one-size-fits-all approach to analyzing COS data.</b></font></p>
<p>This section aims to walk you through taking a brief critical look at your data by taking the first steps of <a class="reference external" href="https://en.wikipedia.org/wiki/Data_binning">binning</a> the data, (to the size of COS’ resolution elements - called <a class="reference external" href="https://hst-docs.stsci.edu/cosihb/chapter-3-description-and-performance-of-the-cos-optics/3-2-size-of-a-resolution-element">resels</a>), and then measuring the Signal-to-Noise Ratio (<a class="reference external" href="https://en.wikipedia.org/wiki/Signal-to-noise_ratio">SNR</a>).</p>
<p><em>Note that the cell below is only necessary if you have chosen to skip Sections #1 and #2</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure these point to your new FUV data</span>
<span class="n">fuv_x1d_filepath</span> <span class="o">=</span> <span class="s1">&#39;./data/mastDownload/HST/lcxv13050&#39;</span>\
                    <span class="s1">&#39;/lcxv13050_x1dsum.fits&#39;</span>
<span class="n">fuv_x1d_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">)</span>

<span class="c1"># Make sure these point to your new NUV data</span>
<span class="n">nuv_x1d_filepath</span> <span class="o">=</span> <span class="s1">&#39;./data/mastDownload/HST/lbbd01020&#39;</span>\
                    <span class="s1">&#39;/lbbd01020_x1dsum.fits&#39;</span>
<span class="n">nuv_x1d_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;count /s /pixel&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
</pre></div>
</div>
</div>
</div>
<p><a id = DataQualV></a></p>
<section id="data-quality-and-initial-cleanup">
<h2>3.1. Data Quality and Initial Cleanup<a class="headerlink" href="#data-quality-and-initial-cleanup" title="Permalink to this heading">#</a></h2>
<p>One of the first things one notices in all of the plots we have made is that all of the segments and stripes have real, useful data, bookended on each side by zeros (see, for example, Fig. 2.1).
These zeros are regions on the active area of the detector which don’t receive and detect light, but <em>are</em> read in as data. We don’t want these datapoints in our spectra, so we will filter them out.</p>
<p>The COS data we have downloaded conveniently have an extension devoted to data quality information. Information on the data quality array and the flags it contains can be found in <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-2-cos-data-files/2-7-error-and-data-quality-arrays">COS Data Handbook</a>.</p>
<p>Let’s first look at the data quality flags. We’ll do this by plotting a histogram, labeling the bins according to what the data quality value of that bin means. We add hatching to the histograms to make it easier to see where certain segments overlap.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To make plotting the bins linearly work,</span>
<span class="c1"># we deal with the base2 logs of the power-of-2 data quality values</span>

<span class="c1"># The values are all powers of two from 2**1 to 2**14</span>
<span class="c1"># The exception is for 0, which is 2**-inf</span>
<span class="n">bins_titles</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">))</span>

<span class="c1"># These are the data quality flags</span>
<span class="n">meanings</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;No Anomalies&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Reed-Solomon Error&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Hot Spot&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Detector Shadow&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Poorly Calibrated Region&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Very-low Response Region&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Background Feature&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Burst&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Pixel Out-of-bounds&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Fill Data&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Pulse Height Out-of-bounds&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Low Response Region&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Bad Time Interval&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Low PHA Feature&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Gain-sag Hole&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Not Used&#39;</span><span class="p">]</span>

<span class="c1"># Pair the meaning with its power of 2 value in a single string</span>
<span class="c1"># One line loop</span>
<span class="n">meanings2</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">bt</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;:&quot;</span><span class="o">+</span><span class="n">mn</span> <span class="k">for</span> <span class="n">bt</span><span class="p">,</span> <span class="n">mn</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">bins_titles</span><span class="p">,</span> <span class="n">meanings</span><span class="p">)]</span>

<span class="c1"># For those less familiar with Python, this works to generate meanings2:</span>
<span class="c1"># meanings2 = [] # Initialize the meanings2 list</span>
<span class="c1"># for i in range(len(meanings)): # Iterate through length of meanings</span>
<span class="c1">#     line = str(bins_titles[i]) + &#39;: &#39; + meanings[i]</span>
<span class="c1">#     meanings2.append(line) # Append line to the meanings2 list</span>

<span class="c1"># Top (FUV) subplot and bottom (NUV) subplot</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>

<span class="c1"># First loop through the 2 FUV segments to make the top plot</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="c1"># Get data quality for current segment</span>
    <span class="n">dataQual</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;DQ&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># All the log2 of the DQ values</span>
    <span class="n">bins_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>
    <span class="c1"># Returning base 2 log of dataQual</span>
    <span class="n">logDQ</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">dataQual</span><span class="p">)</span>
    <span class="c1"># Filling neg infinity values with -1</span>
    <span class="n">logDQ</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">logDQ</span><span class="p">,</span> <span class="n">neginf</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Having there be no x axis ticks for top plot</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

    <span class="c1"># Add hatching to distinguish overlapping colors.</span>
    <span class="c1"># Data, num of bins, color of bin for each segment:</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">logDQ</span><span class="p">,</span>
             <span class="n">bins</span><span class="o">=</span><span class="n">bins_</span><span class="p">,</span>
             <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cmy&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
             <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="o">+</span><span class="mf">0.1</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
             <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
             <span class="c1"># First segment has hatch with vertical lines in bin,</span>
             <span class="c1"># second has horizontal lines</span>
             <span class="n">hatch</span><span class="o">=</span><span class="s1">&#39;|-&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

    <span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
    <span class="c1"># Setting a scale for the y axis</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Fig. 3.1</span><span class="se">\n</span><span class="s2">Data quality flag frequencies&quot;</span><span class="p">,</span>
                  <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="c1"># Loop through 3 NUV stripes to make bottom plot, similar to FUV plot</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
    <span class="c1"># Gather the data</span>
    <span class="n">dataQual</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;DQ&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># All the log2 of the DQ values</span>
    <span class="n">bins_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>

    <span class="c1"># Returning base 2 log of dataQual</span>
    <span class="n">logDQ</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">dataQual</span><span class="p">)</span>
    <span class="n">logDQ</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="p">(</span><span class="n">logDQ</span><span class="p">)</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">)]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="n">ax1</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">logDQ</span><span class="p">,</span>
             <span class="n">bins</span><span class="o">=</span><span class="n">bins_</span><span class="p">,</span>
             <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cmy&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
             <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="mf">0.25</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
             <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">,</span>
             <span class="c1"># Add hatching to distinguish overlapping colors</span>
             <span class="n">hatch</span><span class="o">=</span><span class="s1">&#39;|-/&#39;</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

    <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">bins_</span><span class="o">+</span><span class="mf">0.5</span><span class="p">,</span>
           <span class="c1"># Setting list of x axis tick pos, getting labels ticks</span>
           <span class="n">labels</span><span class="o">=</span><span class="n">meanings2</span><span class="p">,</span>
           <span class="n">rotation</span><span class="o">=</span><span class="s1">&#39;vertical&#39;</span><span class="p">,</span>
           <span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="n">fig</span><span class="o">.</span><span class="n">set_tight_layout</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig3.1.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_2221/2395070808.py:47: RuntimeWarning: divide by zero encountered in log2
  logDQ = np.log2(dataQual)
/tmp/ipykernel_2221/2395070808.py:80: RuntimeWarning: divide by zero encountered in log2
  logDQ = np.log2(dataQual)
</pre></div>
</div>
<img alt="../../../_images/fd0bfe831e5907a5b9f0a89f9568ce74a880543d31fc959454c5a33b19f1f9be.png" src="../../../_images/fd0bfe831e5907a5b9f0a89f9568ce74a880543d31fc959454c5a33b19f1f9be.png" />
</div>
</div>
<p>When the data is processed through <code class="docutils literal notranslate"><span class="pre">CalCOS</span></code>, it has a summed set of “Serious Data Quality” <a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-3-cos-calibration/3-4-descriptions-of-spectroscopic-calibration-steps">SDQ</a> flags attached to it in the form of a single integer between 0 and <span class="math notranslate nohighlight">\(2^{15}\)</span>. This value is found in the 1-th header as <code class="docutils literal notranslate"><span class="pre">&quot;SDQFLAGS&quot;</span></code>. This can be decomposed into the powers of 2 data flags above, (by showing its binary representation,) and tells the pipeline whether to weight that pixel in that exposure by 0 or 1. These weights are then stored in the <code class="docutils literal notranslate"><span class="pre">x1d</span></code> and <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> files, in the <code class="docutils literal notranslate"><span class="pre">&quot;DQ_WGT&quot;</span></code> column. The <code class="docutils literal notranslate"><span class="pre">&quot;DQ_WGT&quot;</span></code> value for a given pixel in the <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> file will then be equal to the number of exposures which went into creating that <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> file in which no serious data quality flags apply to that pixel.</p>
<p>Below, we first see <em>which</em> data quality flags the pipeline considered “serious”, and then demonstrate masking out the datapoints assigned a weight of 0.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdqFlags_fuv</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">fuv_x1d_filepath</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="s1">&#39;SDQFLAGS&#39;</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">&#39;The FUV was processed with SDQFLAGS = </span><span class="si">{</span><span class="n">sdqFlags_fuv</span><span class="si">}</span><span class="s1">, &#39;</span>
    <span class="sa">f</span><span class="s1">&#39;which in binary is </span><span class="si">{</span><span class="nb">bin</span><span class="p">(</span><span class="n">sdqFlags_fuv</span><span class="p">)[</span><span class="mi">2</span><span class="p">:]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Thus, the following DQ-flagged data get weighted by 0:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">char</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">bin</span><span class="p">(</span><span class="n">sdqFlags_fuv</span><span class="p">)[</span><span class="mi">2</span><span class="p">:]):</span>
    <span class="k">if</span> <span class="n">char</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\t</span><span class="si">{</span><span class="p">(</span><span class="mi">2</span><span class="o">**</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">))</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">s place - &#39;</span><span class="p">,</span> <span class="n">meanings</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The FUV was processed with SDQFLAGS = 8346, which in binary is 10000010011010
Thus, the following DQ-flagged data get weighted by 0:

	2&#39;s place -  Hot Spot
	128&#39;s place -  Pixel Out-of-bounds
	1024&#39;s place -  Low Response Region
	2048&#39;s place -  Bad Time Interval
	8192&#39;s place -  Gain-sag Hole
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the datapoints to where there are no serious DQ flags</span>
<span class="n">mask_noSDQ</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;DQ_WGT&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>

<span class="n">wvln_FUVA_noSDQ</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">mask_noSDQ</span><span class="p">]</span>
<span class="n">flux_FUVA_noSDQ</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">mask_noSDQ</span><span class="p">]</span>

<span class="c1"># Make the figure</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Plot the filtered datapoints</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln_FUVA_noSDQ</span><span class="p">,</span> <span class="n">flux_FUVA_noSDQ</span><span class="p">)</span>

<span class="c1"># Format the figure</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Fig 3.2</span><span class="se">\n</span><span class="s2">FUVA without SDQ flagged data&quot;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig3.2.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/f68265dde1a9a3b7a15d627065a380189f8cf8157a75c5f89646316b91a6b5cc.png" src="../../../_images/f68265dde1a9a3b7a15d627065a380189f8cf8157a75c5f89646316b91a6b5cc.png" />
</div>
</div>
<p>Alternatively, we can ignore the <code class="docutils literal notranslate"><span class="pre">SDQFLAGS</span></code> and the weights decided by <code class="docutils literal notranslate"><span class="pre">CalCOS</span></code> and pick which data quality flags we want to filter out. From Fig 3.1, we know the only problem in the FUV and the largest problem in the NUV are pixels flagged by <strong>“out of bounds”</strong>, corresponding to Data Quality: <strong>128</strong>.</p>
<p>We will demonstrate removing only these datapoints from the FUV data below. You could also filter to only data with <em>0: no anomalies</em>, or to any other subset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Creating a plot 12x6 inches with 100 px/in</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># Going through both FUV segments</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="c1"># Getting the data from the current segment&#39;s x1d file</span>
    <span class="n">wvln</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">fluxErr</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;ERROR&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">dataQual</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;DQ&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>

    <span class="c1"># Plotting wavlength vs flux for pixels not out of bounds</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">!=</span> <span class="mf">128.0</span><span class="p">],</span> <span class="n">flux</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">!=</span> <span class="mf">128.0</span><span class="p">],</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
             <span class="n">c</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#d55e00&#39;</span><span class="p">,</span> <span class="s1">&#39;#009e73&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
             <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2"> Cleaned&quot;</span><span class="p">)</span>

    <span class="c1"># Plotting a scatter plot for pixels that are out of bounds</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wvln</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">==</span> <span class="mf">128.0</span><span class="p">],</span> <span class="n">flux</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">==</span> <span class="mf">128.0</span><span class="p">],</span>
                <span class="c1"># FUVA: size 12 black markers</span>
                <span class="c1"># FUVB: size 4 blue markers</span>
                <span class="n">s</span><span class="o">=</span><span class="p">[</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                <span class="n">c</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span>
                <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2"> Pix out-of-bounds&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Fig. 3.3</span><span class="se">\n</span><span class="s2">Removing only out-of-bounds &quot;</span>
          <span class="s2">&quot;pixels from the FUV</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="c1"># Formatting figure and saving</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span> <span class="o">/</span> <span class="s1">&#39;Fig3.3.png&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/238e52b63782fbb628d0fb8d4554770f0a71c8bc565c58e5788164678386f2ee.png" src="../../../_images/238e52b63782fbb628d0fb8d4554770f0a71c8bc565c58e5788164678386f2ee.png" />
</div>
</div>
<section id="exercise-3-1-removing-all-flagged-data-from-the-nuv">
<h3>Exercise 3.1: <em>Removing all flagged data from the NUV</em><a class="headerlink" href="#exercise-3-1-removing-all-flagged-data-from-the-nuv" title="Permalink to this heading">#</a></h3>
<p>The NUV data has flags for more than just out-of-bounds pixels.</p>
<ol class="arabic simple">
<li><p>Filter out all data which has been flagged</p></li>
<li><p>Plot the NUV spectrum from stripes NUVA and NUVB</p></li>
<li><p>Plot the flagged data in another color to see where the removed data is</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Your answer here</span>
</pre></div>
</div>
</div>
</div>
<p><a id=BinningV></a></p>
</section>
</section>
<section id="binning">
<h2>3.2. Binning<a class="headerlink" href="#binning" title="Permalink to this heading">#</a></h2>
<p>Up until now, the data we’ve been working with has one value per “pixel” of the detector.
COS data is generally binned to a resolution element, or <a class="reference external" href="https://hst-docs.stsci.edu/cosihb/chapter-3-description-and-performance-of-the-cos-optics/3-2-size-of-a-resolution-element">resel</a>, when working with it. These resels are <em>generally</em> the following sizes <em>in pixels</em>.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>FUV</p></th>
<th class="head"><p>NUV</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Dispersion Axis</strong></p></td>
<td><p>6</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Cross-Dispersion Axis</strong></p></td>
<td><p>10</p></td>
<td><p>3</p></td>
</tr>
</tbody>
</table>
</div>
<p>So, for the FUV, we generally want to preliminarily bin our <code class="docutils literal notranslate"><span class="pre">x1d</span></code> or <code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> data by 6.</p>
<p><a id=BinfunV></a></p>
<section id="bringing-in-some-useful-functions-for-binning">
<h3>3.2.1. Bringing in Some Useful Functions for Binning<a class="headerlink" href="#bringing-in-some-useful-functions-for-binning" title="Permalink to this heading">#</a></h3>
<p>In the same subdirectory of the <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks">spacetelescope/hst_notebooks GitHub repository</a> as this file (<code class="docutils literal notranslate"><span class="pre">hst_notebooks/COS/ViewData/</span></code>), we also have the <code class="docutils literal notranslate"><span class="pre">Python</span></code> file: <code class="docutils literal notranslate"><span class="pre">cos_functions.py</span></code>, in which we have defined several long functions necessary for this data processing.</p>
<p><em>Note, if you downloaded this Notebook file without cloning the entire GitHub repository, you must also download the file <code class="docutils literal notranslate"><span class="pre">cos_functions.py</span></code> from <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/COS/ViewData">here</a></em>.</p>
<p>In the cell below, we import the function <code class="docutils literal notranslate"><span class="pre">bin_by_resel</span></code> from the file <code class="docutils literal notranslate"><span class="pre">cos_functions.py</span></code>. The <code class="docutils literal notranslate"><span class="pre">bin_by_resel</span></code> function applies the following binning algorithms to an entire COS NUV or FUV dataset. We run the help function to get more information on how the values are binned.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Importing cos_functions.py so we can use it in our Notebook</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cos_functions</span><span class="w"> </span><span class="kn">import</span> <span class="n">bin_by_resel</span>
<span class="n">help</span><span class="p">(</span><span class="n">bin_by_resel</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function bin_by_resel in module cos_functions:

bin_by_resel(data_table, binsize=6, weighted=True, verbose=True)
    Bins an entire COS dataset (in astropy Table form)
        * Wavelength and flux are combined by taking an exposure-time weighted mean of all the pixels in a bin.
        * Errors are combined as above and divided by the square root of the number of pixels in a bin
        * Counts and count rate are summed over all the pixels in a bin.
    
    Parameters:
    data_table (Table) : Astropy Table of COS spectral data.
    binsize (int) : What to bin by.
    weighted (bool) : Whether to weight the averages by exposure time of a pixel; Default is True.
    verbose (bool) : Whether to print major steps the function is taking; Default is True.
    
    Returns:
    Table : New binned table of values
</pre></div>
</div>
</div>
</div>
<p><a id = binfuvV></a></p>
</section>
<section id="binning-the-fuv-data">
<h3>3.2.2. Binning the FUV Data<a class="headerlink" href="#binning-the-fuv-data" title="Permalink to this heading">#</a></h3>
<p><strong>Below we bin our FUV data by 6. Our resulting table is just under <span class="math notranslate nohighlight">\(\dfrac{1}{6}\)</span> the size of our input table, as a few of the last datapoints were cropped.</strong></p>
<p><em>Note</em>, this binning method works well for binsizes which are small compared to the spectrum; however, if you increase your binsize significantly, you may lose important datapoints.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">binned_fuv_data</span> <span class="o">=</span> <span class="n">bin_by_resel</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span>\
    <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">binned_fuv_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Original Dataset Length is &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Binned Dataset Length is &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">binned_fuv_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">fuv_x1d_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;over </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">binned_fuv_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">binned_fuv_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>function `bin_by_resel` is binning by 6

Original Dataset Length is 16384
Binned Dataset Length is 2730
	16384 over 2730 = 6.0014652014652015
</pre></div>
</div>
<div class="output text_html"><div><i>Table length=2</i>
<table id="table140469923758864" class="table-striped table-bordered table-condensed">
<thead><tr><th>EXPTIME</th><th>WAVELENGTH</th><th>FLUX</th><th>ERROR</th><th>ERROR_LOWER</th><th>GROSS</th><th>GCOUNTS</th></tr></thead>
<thead><tr><th>float64</th><th>float64[2730]</th><th>float32[2730]</th><th>float64[2730]</th><th>float64[2730]</th><th>float32[2730]</th><th>float32[2730]</th></tr></thead>
<tr><td>6532.512</td><td>1610.2714602559608 .. 1810.865233806863</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td></tr>
<tr><td>6532.512</td><td>1421.9770427175997 .. 1622.516178750543</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td><td>0.0 .. 0.0</td></tr>
</table></div></div></div>
</div>
<p><a id=SnrV></a></p>
</section>
</section>
<section id="estimating-the-signal-to-noise-snr-ratio">
<h2>3.3. Estimating the Signal-to-Noise (SNR) Ratio<a class="headerlink" href="#estimating-the-signal-to-noise-snr-ratio" title="Permalink to this heading">#</a></h2>
<p><a id=snrfunV></a></p>
<section id="bringing-in-a-useful-function-for-estimating-snr">
<h3>3.3.1.  Bringing in a Useful Function for Estimating SNR<a class="headerlink" href="#bringing-in-a-useful-function-for-estimating-snr" title="Permalink to this heading">#</a></h3>
<p>Below, we define a function to estimate the Signal-to-Noise ratio (SNR or S/N) of our data. One way to estimate the SNR for photon-counting instruments like COS is by the square-root of their number of gross counts. This is a simplification of the equation given for point sources in the <a class="reference external" href="https://hst-docs.stsci.edu/cosihb/chapter-7-exposure-time-calculator-etc/7-3-sensitivity-count-rate-and-signal-to-noise">COS Instrument Handbook Section 7.3</a>:</p>
<div class="math notranslate nohighlight">
\[\dfrac{Signal}{Noise} = \dfrac{C \times t}{\sqrt{C \times t + N_{pix} \times (B_{sky} + B_{det}) \times t}}\]</div>
<p>For bright point sources, <span class="math notranslate nohighlight">\(C &gt;&gt; B_{sky} + B_{det}\)</span>, and this approaches the simpler:</p>
<div class="math notranslate nohighlight">
\[\dfrac{Signal}{Noise} = \sqrt{C \times t} = \sqrt{Gross\ Counts}\]</div>
<p><font size="5">Caution!</font></p>
<img src=figures/warning.png width ="60" title="CAUTION!"> 
This simplification may not hold very well if your source is diffuse or faint.<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cos_functions</span><span class="w"> </span><span class="kn">import</span> <span class="n">estimate_snr</span>
<span class="n">help</span><span class="p">(</span><span class="n">estimate_snr</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function estimate_snr in module cos_functions:

estimate_snr(data_table, snr_range=[-1, -1], bin_data_first=False, binsize_=6, weighted=False, verbose=True)
    Gets an estimate of the Signal to Noise Ratio (SNR), either over wvln-range or whole spectrum, using Poisson noise assumption SNR ~sqrt(N_Counts).
    Weights the SNR
    
    Parameters:
    data_table (Astropy Table) : astropy table of COS data.
    snr_range (list) : list of two values - [wvln_range_start , wvln_range_end]; Default is [-1,-1], indicating that we will take over all values.
    bin_data_first (bool) : Should we begin by binning the data by the binsize_? Default is False.
    binsize_ (int) : If bin_data_first == True, what to bin by; Default is 6 for fuv resel.
    weighted (bool) : Do you want the average to be an exposure time weighted average rather than the default unweighted mean; Default is False.
    verbose (bool) : Whether to give a few print statements; default is True.
    
    Returns:
    float : A single value for the exptime-weighted average or mean SNR over the specified snr_range; -1 if no specified range.
    nested list : 1st level of list corresponds to the segments/rows of the input data_table, 2nd level holds wvln, snr, segmentnumber over snr_range -
                 ie [[-1,-1,-1],[wvln over range array, wvln over range array, row in input data_table int][-1,-1,-1]].
</pre></div>
</div>
</div>
</div>
<p><strong>The function <code class="docutils literal notranslate"><span class="pre">estimate_snr</span></code> gives a <span class="math notranslate nohighlight">\(\sqrt{Gross\ Counts}\)</span> estimate of SNR for an input <code class="docutils literal notranslate"><span class="pre">Astropy</span></code> Table of COS data.</strong></p>
<p>By default, it does not bin the data before calculating the SNR, nor does it weight the output SNR by exposure time; however, it has the functionality to do both. In most cases, a range should be specified over which to calculate the SNR. If no range is specified, then SNR will be computed for all Resels, but no mean/weighted average will be computed; instead, -1 will be returned in place of the mean.</p>
<p><a id = snrregV></a></p>
</section>
<section id="choosing-a-region-to-calculate-snr">
<h3>3.3.2. Choosing a Region to calculate SNR<a class="headerlink" href="#choosing-a-region-to-calculate-snr" title="Permalink to this heading">#</a></h3>
<p>The SNR should be calculated over a region of continuum, rather than one with significant spectral lines.
Looking back to <span class="xref myst">Fig. 2.4</span>, we see a broad stretch with no obvious lines from [1675 - 1690 Å] in the FUVA segment. We’ll make a quick plot of that region below just to check.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>  <span class="c1"># Set up figure</span>
<span class="c1"># Select data</span>
<span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">,</span> <span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="c1"># Create mask for data</span>
<span class="n">wvln_range_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&gt;</span> <span class="mi">1670</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&lt;</span> <span class="mi">1700</span><span class="p">)</span>

<span class="c1"># Generate the plot itself</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">[</span><span class="n">wvln_range_mask</span><span class="p">],</span> <span class="n">flux</span><span class="p">[</span><span class="n">wvln_range_mask</span><span class="p">])</span>

<span class="c1"># Add lines showing the range we&#39;ll calculate SNR over</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">1675</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">1690</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Format the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Fig. 3.4</span><span class="se">\n</span><span class="s2">Range for SNR Calculation&quot;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>

<span class="c1"># Formatting figure and then saving to plots directory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span><span class="o">/</span><span class="s1">&#39;Fig3.4.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/605ec0620e4e16d1b61b4e9fd56a8533e271640b5a8e71e7938b106150172522.png" src="../../../_images/605ec0620e4e16d1b61b4e9fd56a8533e271640b5a8e71e7938b106150172522.png" />
</div>
</div>
<p>Our chosen region, between the black lines, looks quite good for calculating SNR, as it lacks obvious spectral lines.
<a id = snrcalcV></a></p>
</section>
<section id="calculating-the-snr">
<h3>3.3.3. Calculating the SNR<a class="headerlink" href="#calculating-the-snr" title="Permalink to this heading">#</a></h3>
<p>Below, we calculate the SNR of the FUV data over the region [1675, 1690]. <em>Note</em>, we input the unbinned data, and bin by an FUV Resel of 6 first. We’ll leave <code class="docutils literal notranslate"><span class="pre">verbose</span> <span class="pre">=</span> <span class="pre">True</span></code>, so we can see what is going on in the function.</p>
<p>It turns out this dataset has a fairly low SNR. After calculating the SNR and mean SNR over the specified region, we plot both over wavelength.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the SNR</span>
<span class="c1"># Set the correct resel size as in the data handbook</span>
<span class="n">res_size</span> <span class="o">=</span> <span class="mi">6</span>

<span class="c1"># Getting SNR over wavelengths of 1675-1690 with data binned by 6 res</span>
<span class="n">meanSNR_range1675_1690</span><span class="p">,</span> <span class="n">range1675_1690</span> <span class="o">=</span> <span class="n">estimate_snr</span><span class="p">(</span>
                                            <span class="n">fuv_x1d_data</span><span class="p">,</span>
                                            <span class="n">bin_data_first</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                            <span class="n">binsize_</span><span class="o">=</span><span class="n">res_size</span><span class="p">,</span>
                                            <span class="n">snr_range</span><span class="o">=</span><span class="p">[</span><span class="mi">1675</span><span class="p">,</span> <span class="mi">1690</span><span class="p">])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="n">wvln_over_range</span> <span class="o">=</span> <span class="n">range1675_1690</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">snr_over_range</span> <span class="o">=</span> <span class="n">range1675_1690</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># range1675_1690 contains a list of data for each</span>
<span class="c1"># segment of the detector. If the wavelength range</span>
<span class="c1"># for calculating S/N is in FUVA, you would</span>
<span class="c1"># access the WAVELENGTH and S/N as:</span>
<span class="c1">#     range1675_1690[1][0], range1675_1690[1][1]</span>

<span class="c1"># Plotting the SNR by wavelength in range 1675 - 1690 angstroms</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln_over_range</span><span class="p">,</span> <span class="n">snr_over_range</span><span class="p">,</span>
         <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#466599&#39;</span><span class="p">,</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
         <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
         <span class="n">markerfacecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
         <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;SNR Binned by Resel size </span><span class="si">{</span><span class="n">res_size</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">meanSNR_range1675_1690</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-.&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
            <span class="c1"># Plot a horizontal line at y = mean SNR</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Mean SNR over the Region&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Fig. 3.5</span><span class="se">\n</span><span class="s2">Results of SNR Calculation by Wavelength&quot;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;SNR&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">plotsdir</span><span class="o">/</span><span class="s1">&#39;Fig3.5.png&#39;</span><span class="p">),</span>
            <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>First, Binning the data by  6
function `bin_by_resel` is binning by 6
In range on 0-th segment with limits: 1610.2714602559608 1810.865233806863 
Unweighted mean SNR over the range [1675, 1690] is: 8.130635261535645
Out of range on 1-th segment with limits: 1421.9770427175997 1622.516178750543
</pre></div>
</div>
<img alt="../../../_images/beaaf160812f87e8080f305ff83c7df8509bda0daf12099a352b6814511543d8.png" src="../../../_images/beaaf160812f87e8080f305ff83c7df8509bda0daf12099a352b6814511543d8.png" />
</div>
</div>
</section>
<section id="exercise-3-2-binning-the-nuv-data-and-calculating-its-snr">
<h3>Exercise 3.2. <em>Binning the NUV Data and Calculating its SNR</em><a class="headerlink" href="#exercise-3-2-binning-the-nuv-data-and-calculating-its-snr" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Bin the NUV data, remembering that the resel size is different than that for the FUV</p></li>
<li><p>Find a continuum region of the NUV data with no significant spectral lines</p></li>
<li><p>Calculate the mean SNR over that region</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Your code here</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="congratulations-you-finished-this-notebook">
<h2>Congratulations! You finished this Notebook!<a class="headerlink" href="#congratulations-you-finished-this-notebook" title="Permalink to this heading">#</a></h2>
<section id="there-are-more-cos-data-walkthrough-notebooks-on-different-topics-you-can-find-them-here">
<h3>There are more COS data walkthrough Notebooks on different topics. You can find them <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/COS">here</a>.<a class="headerlink" href="#there-are-more-cos-data-walkthrough-notebooks-on-different-topics-you-can-find-them-here" title="Permalink to this heading">#</a></h3>
</section>
</section>
<hr class="docutils" />
<section id="about-this-notebook">
<h2>About this Notebook<a class="headerlink" href="#about-this-notebook" title="Permalink to this heading">#</a></h2>
<p><strong>Author:</strong> <span class="xref myst">Nat Kerman</span></p>
<p><strong>Updated On:</strong> 2023-03-28</p>
<blockquote>
<div><p><em>This tutorial was generated to be in compliance with the <a class="reference external" href="https://github.com/spacetelescope/style-guides">STScI style guides</a> and would like to cite the <a class="reference external" href="https://github.com/spacetelescope/style-guides/blob/master/templates/example_notebook.ipynb">Jupyter guide</a> in particular.</em></p>
</div></blockquote>
</section>
<section id="citations">
<h2>Citations<a class="headerlink" href="#citations" title="Permalink to this heading">#</a></h2>
<p>If you use <code class="docutils literal notranslate"><span class="pre">astropy</span></code>, <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code>, <code class="docutils literal notranslate"><span class="pre">astroquery</span></code>, or <code class="docutils literal notranslate"><span class="pre">numpy</span></code> for published research, please cite the
authors. Follow these links for more information about citations:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://docs.astropy.org/en/stable/index.html">Citing <code class="docutils literal notranslate"><span class="pre">astropy</span></code></a></p></li>
<li><p><a class="reference external" href="https://numpy.org/citing-numpy/">Citing <code class="docutils literal notranslate"><span class="pre">numpy</span></code></a></p></li>
<li><p><a class="reference external" href="https://matplotlib.org/">Citing <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code></a></p></li>
<li><p><a class="reference external" href="https://astroquery.readthedocs.io/en/latest/">Citing <code class="docutils literal notranslate"><span class="pre">astroquery</span></code></a></p></li>
</ul>
<hr class="docutils" />
<p><span class="xref myst">Top of Page</span>
<a class="reference internal" href="https://raw.githubusercontent.com/spacetelescope/notebooks/master/assets/stsci_pri_combo_mark_horizonal_white_bkgd.png"><img alt="Space Telescope Logo" src="https://raw.githubusercontent.com/spacetelescope/notebooks/master/assets/stsci_pri_combo_mark_horizonal_white_bkgd.png" style="width: 200px;" /></a></p>
</section>
<section id="exercise-solutions">
<h2>Exercise Solutions:<a class="headerlink" href="#exercise-solutions" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Exercise 1.1 Solution:</span>

<span class="c1"># 1</span>
<span class="c1"># Here we read in the data - we may ignore slash warnings</span>
<span class="n">nuv_x1d_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span>

<span class="c1"># 2</span>
<span class="n">nuv_x1d_header</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">)</span>
<span class="n">nuv_x1d_header1</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">getheader</span><span class="p">(</span><span class="n">nuv_x1d_filepath</span><span class="p">,</span> <span class="n">ext</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Processed at MJD = </span><span class="si">{</span><span class="n">nuv_x1d_header</span><span class="p">[</span><span class="s1">&#39;PROCTIME&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Taken with </span><span class="si">{</span><span class="n">nuv_x1d_header1</span><span class="p">[</span><span class="s1">&#39;NUMFLASH&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="s2">&quot; wavecal flashes</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># 3</span>
<span class="n">nuv_asn_data</span> <span class="o">=</span> <span class="n">Table</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">nuv_asn_filepath</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nuv_asn_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Processed at MJD = 60592.68001157
Taken with 1 wavecal flashes

   MEMNAME        MEMTYPE     MEMPRSNT
-------------- -------------- --------
     LBBD01HPQ         EXP-FP        1
     LBBD01020        PROD-FP        1
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;erg /s /cm**2 /angstrom&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: &#39;count /s /pixel&#39; contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Exercise 2.1 Solution:</span>

<span class="c1"># [0] Accesses FUVA, the longer wvln segment.</span>
<span class="c1"># Gets data needed to plot a spectrum</span>
<span class="n">wvln</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># [1] Accesses FUVB</span>
<span class="c1"># Adds unit to the wvln of Angstroms</span>
<span class="n">wvln</span> <span class="o">*=</span> <span class="n">u</span><span class="o">.</span><span class="n">AA</span>
<span class="c1"># Adds unit to the flux of erg / (Angstrom cm2 s)</span>
<span class="n">flux</span> <span class="o">*=</span> <span class="n">u</span><span class="o">.</span><span class="n">erg</span><span class="o">/</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">second</span> <span class="o">*</span> <span class="n">u</span><span class="o">.</span><span class="n">AA</span> <span class="o">*</span> <span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">cm</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>
<span class="n">flux</span> <span class="o">/=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">flux</span><span class="p">))</span>
<span class="c1"># Allows us to view the units attached to the data as the axes labels.</span>
<span class="c1"># This isn&#39;t generally necessary unless you want the unit support.</span>
<span class="k">with</span> <span class="n">quantity_support</span><span class="p">():</span>
    <span class="n">fig1</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                            <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

    <span class="c1"># The next few lines are where we place data onto the plot:</span>
    <span class="c1">###############</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="n">segment</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Exercise 2.1</span><span class="se">\n</span><span class="s2">Simple COS Segment FUVB Spectrum&quot;</span><span class="p">,</span>
                 <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

    <span class="c1"># Setting the x-axis label</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="c1"># Setting the y-axis label</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">1420</span><span class="p">,</span> <span class="mi">1620</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/11facca0b7bb24db6aedc299da9ad1377c73ef05ee3a4d1fdb567558de0db816.png" src="../../../_images/11facca0b7bb24db6aedc299da9ad1377c73ef05ee3a4d1fdb567558de0db816.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Exercise 2.2 Solution</span>

<span class="c1"># Selects all useful data for the chosen segment = FUVA</span>
<span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;WAVELENGTH&#39;</span><span class="p">,</span> <span class="s1">&#39;FLUX&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">fluxErr</span><span class="p">,</span> <span class="n">segment</span> <span class="o">=</span> <span class="n">fuv_x1d_data</span><span class="p">[</span><span class="s1">&#39;ERROR&#39;</span><span class="p">,</span> <span class="s1">&#39;SEGMENT&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="n">minx</span><span class="p">,</span> <span class="n">maxx</span> <span class="o">=</span> <span class="mi">1635</span><span class="p">,</span> <span class="mi">1675</span>
<span class="n">miny</span><span class="p">,</span> <span class="n">maxy</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">flux</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">flux</span><span class="p">)</span>

<span class="n">wvln_range_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&gt;</span> <span class="n">minx</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">wvln</span> <span class="o">&lt;</span> <span class="n">maxx</span><span class="p">)</span>

<span class="n">wvln</span> <span class="o">=</span> <span class="n">wvln</span><span class="p">[</span><span class="n">wvln_range_mask</span><span class="p">]</span>
<span class="n">flux</span> <span class="o">=</span> <span class="n">flux</span><span class="p">[</span><span class="n">wvln_range_mask</span><span class="p">]</span>
<span class="n">fluxErr</span> <span class="o">=</span> <span class="n">fluxErr</span><span class="p">[</span><span class="n">wvln_range_mask</span><span class="p">]</span>

<span class="n">rangex</span> <span class="o">=</span> <span class="n">maxx</span> <span class="o">-</span> <span class="n">minx</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="n">nRows</span> <span class="o">=</span> <span class="mi">3</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nRows</span><span class="p">):</span>
    <span class="n">min_</span> <span class="o">=</span> <span class="n">minx</span> <span class="o">+</span> <span class="n">i</span><span class="o">*</span><span class="n">rangex</span><span class="o">/</span><span class="n">nRows</span>
    <span class="n">max_</span> <span class="o">=</span> <span class="n">minx</span> <span class="o">+</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">rangex</span><span class="o">/</span><span class="n">nRows</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">nRows</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># A way to set title, xlabel, and ylabel independent of # of rows</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Ex 2.2 Solution FUV</span><span class="si">{</span><span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Segment </span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="s2">&quot;spectrum split into segments&quot;</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">nRows</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
                      <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">nRows</span><span class="o">/</span><span class="mi">2</span><span class="p">):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA ^{-1}$]&#39;</span><span class="p">,</span>
                      <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">errorbar</span><span class="p">(</span><span class="n">wvln</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span> <span class="n">fluxErr</span><span class="p">,</span>
                <span class="n">c</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">rainbow</span><span class="p">((</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="n">nRows</span><span class="p">),</span>
                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
                <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
                <span class="n">markerfacecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span>
                <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">mew</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_</span><span class="p">,</span> <span class="n">max_</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/a318e5f33f142e578ac553159b34d780aab9944dfddd83ab7676542201428e4d.png" src="../../../_images/a318e5f33f142e578ac553159b34d780aab9944dfddd83ab7676542201428e4d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Exercise 3.1 Solution</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">dataQual</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;DQ&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">wvln</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">fluxErr</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;ERROR&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">nuv_x1d_data</span><span class="p">[</span><span class="s2">&quot;SEGMENT&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wvln</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">==</span> <span class="mi">0</span><span class="p">],</span> <span class="n">flux</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">==</span> <span class="mi">0</span><span class="p">],</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
             <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2"> Un-Flagged Pixel&quot;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wvln</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">],</span> <span class="n">flux</span><span class="p">[</span><span class="n">dataQual</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">],</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">c</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span>
                <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2"> Flagged Pixel&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span>
    <span class="s2">&quot;Ex 3.1 Solution</span><span class="se">\n</span><span class="s2">Removing only out-of-bounds &quot;</span>
    <span class="s2">&quot;pixels from the NUV</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/491e6e6d33f69873f5c52db0fa55e1d70c58416beed5fbdbe67b6a65e58868a7.png" src="../../../_images/491e6e6d33f69873f5c52db0fa55e1d70c58416beed5fbdbe67b6a65e58868a7.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Exercise 3.2 Solution</span>

<span class="c1"># This whole process can be accomplished in one line, provided we</span>
<span class="c1"># use Fig 2.6 to find a line free region. We use the estimate_snr</span>
<span class="c1"># function to bin by NUV resel size of 3, then mean calculate</span>
<span class="c1"># SNR over the region.</span>

<span class="n">meanSNR_nuv_range</span><span class="p">,</span> <span class="n">nuv_range</span> <span class="o">=</span> <span class="n">estimate_snr</span><span class="p">(</span>
                                    <span class="n">nuv_x1d_data</span><span class="p">,</span>
                                    <span class="n">bin_data_first</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                    <span class="n">binsize_</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                                    <span class="n">snr_range</span><span class="o">=</span><span class="p">[</span><span class="mi">3400</span><span class="p">,</span> <span class="mi">3450</span><span class="p">])</span>

<span class="c1"># Let&#39;s plot it as in Fig 3.5:</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">nuv_range</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">nuv_range</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;SNR Binned by Resel&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">meanSNR_nuv_range</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-.&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Mean SNR over the Region&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Ex. 3.2 Solution</span><span class="se">\n</span><span class="s2">Results of SNR Calculation&quot;</span><span class="p">,</span>
          <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;SNR&#39;</span><span class="p">,</span>
           <span class="n">size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>First, Binning the data by  3
function `bin_by_resel` is binning by 3
Out of range on 0-th segment with limits: 2055.2227946721123 2551.202543205818
In range on 1-th segment with limits: 3144.4680068817165 3639.1168209501548 
Unweighted mean SNR over the range [3400, 3450] is: 29.33061408996582
Out of range on 2-th segment with limits: 2141.381207452109 2387.0087020695464
</pre></div>
</div>
<img alt="../../../_images/3c857a52a1c192610dee9118000e4e93784d1600938b8f2972d64109506e6f93.png" src="../../../_images/3c857a52a1c192610dee9118000e4e93784d1600938b8f2972d64109506e6f93.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "spacetelescope/hst_notebooks",
            ref: "main",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks/COS/ViewData"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../DataDl/DataDl.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Downloading COS Data</p>
      </div>
    </a>
    <a class="right-next"
       href="../AsnFile/AsnFile.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Modifying or Creating an Association File</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Viewing COS Data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning Goals</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">0. Introduction</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-one-cell-summary-of-this-notebook-s-key-points">0.1. A one-cell summary of this Notebook’s key points:</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#now-returning-to-our-more-detailed-walkthrough">Now, returning to our more detailed walkthrough</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#notes-for-those-new-to-python-jupyter-coding">Notes for those new to <code class="docutils literal notranslate"><span class="pre">Python</span></code>/Jupyter/Coding:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-in-the-data">1. Reading in the data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#investigating-the-data-basics">1.1. Investigating the Data - <em>Basics</em></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-in-the-x1d-x1dsum-main-data">1.2. Reading in the <code class="docutils literal notranslate"><span class="pre">x1d</span></code>/<code class="docutils literal notranslate"><span class="pre">x1dsum</span></code> Main Data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fig-1-1-from-cos-dhb-fig-1-6">Fig. 1.1 from COS DHB Fig. 1.6</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fig-1-2-from-cos-dhb-fig-1-10">Fig. 1.2 from COS DHB Fig. 1.10</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-association-asn-file">1.3. The Association (<code class="docutils literal notranslate"><span class="pre">asn</span></code>) file</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1-1-finding-metadata-for-the-nuv">Exercise 1.1. <em>Finding Metadata for the NUV</em></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-data">2. Plotting the Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-an-fuv-spectrum">2.1 Plotting an FUV Spectrum</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-first-plot">2.1.1 Our First Plot</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-1-a-clearer-plot">Exercise 2.1: A clearer plot</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-complex-look-at-the-entire-fuv">2.1.2. A Complex Look at the Entire FUV</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-closer-at-parts-of-the-fuv-spectrum">2.1.3. Looking Closer at Parts of the FUV Spectrum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-2-choosing-and-plotting-a-useful-data-range">Exercise 2.2 Choosing and plotting a useful data range</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-plotting-the-data-with-specutils-optional">2.1.4. Reading and plotting the data with <code class="docutils literal notranslate"><span class="pre">specutils</span></code> (<em>OPTIONAL</em>)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-an-nuv-spectrum">2.2. Plotting an NUV Spectrum</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examining-the-first-order-spectrum">2.2.1. Examining the first-order spectrum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examining-the-second-order-spectrum">2.2.2. Examining the second-order spectrum</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#preliminary-review-of-the-data">3. Preliminary Review of the Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-quality-and-initial-cleanup">3.1. Data Quality and Initial Cleanup</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-1-removing-all-flagged-data-from-the-nuv">Exercise 3.1: <em>Removing all flagged data from the NUV</em></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#binning">3.2. Binning</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bringing-in-some-useful-functions-for-binning">3.2.1. Bringing in Some Useful Functions for Binning</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#binning-the-fuv-data">3.2.2. Binning the FUV Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#estimating-the-signal-to-noise-snr-ratio">3.3. Estimating the Signal-to-Noise (SNR) Ratio</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bringing-in-a-useful-function-for-estimating-snr">3.3.1.  Bringing in a Useful Function for Estimating SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-a-region-to-calculate-snr">3.3.2. Choosing a Region to calculate SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-the-snr">3.3.3. Calculating the SNR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-2-binning-the-nuv-data-and-calculating-its-snr">Exercise 3.2. <em>Binning the NUV Data and Calculating its SNR</em></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#congratulations-you-finished-this-notebook">Congratulations! You finished this Notebook!</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#there-are-more-cos-data-walkthrough-notebooks-on-different-topics-you-can-find-them-here">There are more COS data walkthrough Notebooks on different topics. You can find them here.</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#about-this-notebook">About this Notebook</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citations">Citations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-solutions">Exercise Solutions:</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By STScI
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022-2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>